<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.4.533">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">


<title>Background</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="ai-ontology-entity-linking-for-clinical-content_files/libs/clipboard/clipboard.min.js"></script>
<script src="ai-ontology-entity-linking-for-clinical-content_files/libs/quarto-html/quarto.js"></script>
<script src="ai-ontology-entity-linking-for-clinical-content_files/libs/quarto-html/popper.min.js"></script>
<script src="ai-ontology-entity-linking-for-clinical-content_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="ai-ontology-entity-linking-for-clinical-content_files/libs/quarto-html/anchor.min.js"></script>
<link href="ai-ontology-entity-linking-for-clinical-content_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="ai-ontology-entity-linking-for-clinical-content_files/libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="ai-ontology-entity-linking-for-clinical-content_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="ai-ontology-entity-linking-for-clinical-content_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="ai-ontology-entity-linking-for-clinical-content_files/libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">


</head>

<body class="fullcontent">

<div id="quarto-content" class="page-columns page-rows-contents page-layout-article">

<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Background</h1>
</div>



<div class="quarto-title-meta">

    
  
    
  </div>
  


</header>


<section id="background" class="level1">
<h1>Background</h1>
<section id="ontology-driven-entity-linking-in-biomedical-clinical-text-using-plain-language-triplestore" class="level2">
<h2 class="anchored" data-anchor-id="ontology-driven-entity-linking-in-biomedical-clinical-text-using-plain-language-triplestore">Ontology-Driven Entity Linking in Biomedical / Clinical Text using Plain Language “Triplestore”</h2>
<hr>
<p>Jung Hoon Son, M.D.</p>
<div class="img-fluid quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="ai-ontology-entity-linking-for-clinical-content_files/figure-html/cell-1-1-e42c5e98-7636-4c74-83dd-6036de865d88.png" class="img-fluid figure-img"></p>
<figcaption>image.png</figcaption>
</figure>
</div>
</section>
<section id="sentencestore-triplestore-as-plain-language" class="level2">
<h2 class="anchored" data-anchor-id="sentencestore-triplestore-as-plain-language">SentenceSTORE: “Triplestore” as Plain Language</h2>
<blockquote class="blockquote">
<p>Semantic Triplestore Organization &amp; Reasoning Engine (STORE)</p>
</blockquote>
<p>This is my best attempt at a catchy name that mimics the concept of triplestores used in ontologies into plain language (English), because I realize LLM work at a many-to-many mapping possible, but it understands the rules of language the best.</p>
<p>For those unfamiliar “triple store” is a general description to store “triples” (https://en.wikipedia.org/wiki/Semantic_triple) often used in knowledge representation, particularly used with ontologies and semantic web. If you are unfamiliar with RDF/XML, OWL, Turtle, SPARQL and never wanted to approach them, this might be a solution for you in at least leveraging ontologies.</p>
<p>Consider this an attempt to reimagine the concept of triplestores for the LLM era. While traditional triplestores store knowledge in rigid <strong>[subject]-[predicate]-[object]</strong> formats, unoriginally named <strong>SentenceSTORE</strong> uses natural language sentences as its fundamental unit of storage.</p>
<hr>
</section>
<section id="terminology-code-mapping-task-has-a-hallucination-problem" class="level2">
<h2 class="anchored" data-anchor-id="terminology-code-mapping-task-has-a-hallucination-problem">Terminology code mapping task has a hallucination problem</h2>
<p>It’s well known to the healthcare world that chatbots are terrible for medical coding tasks. Personally, I totally get it, <strong>given the tokenization techniques, it’s tough to make sense of these alphanumeric patterns</strong> from the standpoint of LLM training.</p>
<p>Example, recent NEJM (New Englang Journal of Medicine) AI article: https://ai.nejm.org/doi/full/10.1056/AIdbp2300040</p>
</section>
<section id="gemini-pro-1.5-base-model-poor" class="level2">
<h2 class="anchored" data-anchor-id="gemini-pro-1.5-base-model-poor">Gemini Pro 1.5 Base Model: Poor</h2>
<p>I’ve tested the Gemini Pro 1.5, and generally it’s in line with the NEJM paper. Medical terminology extraction / medical coding is not a native LLM job.</p>
<p>But overall, ontologies have a lot of accrued, carefully curated knowledge that most people are open-access to.</p>
<hr>
</section>
</section>
<section id="ontology-basics" class="level1">
<h1>Ontology Basics</h1>
<section id="intro-to-owlready2" class="level2">
<h2 class="anchored" data-anchor-id="intro-to-owlready2">Intro to <strong>Owlready2</strong></h2>
<p>Let me demonstrate why ontologies are hard to use. Using one of the easier (maybe only) libraries in python that allows us to access data in an ontology, and note that this is <strong>very rudimentary exploration of ontologies</strong>.</p>
<p>We will be using one of the few native-python libraries for dealing with ontologies to extract the triple stores in the OWL files we will be using, <code>owlready2</code> - https://owlready2.readthedocs.io/en/latest/</p>
<section id="step-1-install-owlready2" class="level3">
<h3 class="anchored" data-anchor-id="step-1-install-owlready2">Step 1: Install owlready2</h3>
<div class="sourceCode" id="cb1"><pre class="sourceCode bash code-with-copy"><code class="sourceCode bash"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="ex">pip</span> install owlready2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</section>
<section id="step-2-download-owl-files" class="level3">
<h3 class="anchored" data-anchor-id="step-2-download-owl-files">Step 2: Download OWL files</h3>
<ul>
<li>Human Phenotype Ontology (https://hpo.jax.org/): [<a href="https://arc.net/l/quote/ozqvisbu">.owl File</a>]</li>
<li>Mondo Disease Ontology (https://mondo.monarchinitiative.org/) [<a href="http://purl.obolibrary.org/obo/mondo.owl">.owl File</a>]</li>
<li>Disease Ontology (https://disease-ontology.org/do) [<a href="https://github.com/DiseaseOntology/HumanDiseaseOntology/raw/refs/heads/main/src/ontology/doid-merged.owl">.owl File</a>]</li>
</ul>
</section>
<section id="step-3-follow-along-if-you-wish-to-get-an-intro-to-ontologies" class="level3">
<h3 class="anchored" data-anchor-id="step-3-follow-along-if-you-wish-to-get-an-intro-to-ontologies">Step 3: Follow along if you wish to get an intro to ontologies</h3>
<hr>
</section>
</section>
<section id="querying-ontologies" class="level2">
<h2 class="anchored" data-anchor-id="querying-ontologies">Querying Ontologies</h2>
<div id="cell-9" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:20:06.069428Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:20:06.068313Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:20:50.152738Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:20:50.151452Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:20:06.069373Z&quot;}" data-trusted="true" data-execution_count="1">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="op">!</span>pip install owlready2</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>Requirement already satisfied: owlready2 in /Users/jung/.pyenv/versions/3.12.6/envs/llm-3.12.6/lib/python3.12/site-packages (0.46)

[notice] A new release of pip is available: 24.2 -&gt; 24.3.1
[notice] To update, run: pip install --upgrade pip</code></pre>
</div>
</div>
<section id="finding-a-concept-with-name-leukemia" class="level3">
<h3 class="anchored" data-anchor-id="finding-a-concept-with-name-leukemia">Finding a concept with name “leukemia”</h3>
<div id="cell-11" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:20:50.155242Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:20:50.154898Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:21:11.170596Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:21:11.169416Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:20:50.155209Z&quot;}" data-trusted="true" data-execution_count="3">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> owlready2 <span class="im">import</span> get_ontology, get_namespace</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>onto <span class="op">=</span> get_ontology(<span class="st">"ontologies/mondo.owl"</span>).load()</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>onto_hp<span class="op">=</span> get_ontology(<span class="st">"ontologies/hp.owl"</span>).load()</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>onto_do<span class="op">=</span> get_ontology(<span class="st">"ontologies/doid-merged.owl"</span>).load()</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>obo <span class="op">=</span> get_namespace(<span class="st">"http://purl.obolibrary.org/obo/"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-12" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:21:11.172212Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:21:11.171892Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:21:11.181545Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:21:11.180744Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:21:11.172182Z&quot;}" data-trusted="true" data-execution_count="4">
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>search_result <span class="op">=</span> onto.search(label<span class="op">=</span><span class="st">"leukemia"</span>)</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>entity <span class="op">=</span> search_result[<span class="dv">0</span>]</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>entity</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="4">
<pre><code>obo.MONDO_0005059</code></pre>
</div>
</div>
</section>
<section id="get-the-label-of-entity" class="level3">
<h3 class="anchored" data-anchor-id="get-the-label-of-entity">Get the label of entity</h3>
<div id="cell-14" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:21:11.183743Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:21:11.183415Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:21:11.223684Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:21:11.222689Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:21:11.183688Z&quot;}" data-trusted="true" data-execution_count="5">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>entity.label</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="5">
<pre><code>['leukemia']</code></pre>
</div>
</div>
</section>
<section id="get-external-terminologies-ontology-ids" class="level3">
<h3 class="anchored" data-anchor-id="get-external-terminologies-ontology-ids">Get external terminologies / ontology IDs</h3>
<ul>
<li>You can get SNOMEDCT, ICD10, UMLS, and various other ontology IDs here.</li>
</ul>
<div id="cell-16" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:21:11.225244Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:21:11.224915Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:21:11.238023Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:21:11.236990Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:21:11.225206Z&quot;}" data-trusted="true" data-execution_count="6">
<div class="sourceCode cell-code" id="cb9"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>entity.hasDbXref</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="6">
<pre><code>['DOID:1240',
 'EFO:0000565',
 'HP:0001909',
 'ICD9:207',
 'ICD9:207.8',
 'ICD9:207.80',
 'ICD9:208',
 'ICD9:208.8',
 'ICD9:208.80',
 'ICD9:208.9',
 'ICD9:208.90',
 'ICDO:9800/3',
 'MEDGEN:9725',
 'MESH:D007938',
 'NANDO:2100002',
 'NCIT:C3161',
 'SCTID:93143009',
 'UMLS:C0023418']</code></pre>
</div>
</div>
</section>
<section id="get-parent-concept-of-an-entity" class="level3">
<h3 class="anchored" data-anchor-id="get-parent-concept-of-an-entity">Get “parent” concept of an entity</h3>
<p>Note duplication - happens with multiple ontologies referencing same entity (HPO and MONDO has some duplications) and Owlready2’s implementation.</p>
<p>Not a big deal for us, since we’ll just use the unique relationships.</p>
<div id="cell-18" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:21:11.239664Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:21:11.239291Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:21:11.254784Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:21:11.253771Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:21:11.239604Z&quot;}" data-trusted="true" data-execution_count="7">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a>[x.label <span class="cf">for</span> x <span class="kw">in</span> entity.ancestors()]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="7">
<pre><code>[['hematopoietic and lymphoid cell neoplasm'],
 [locstr('specifically dependent continuant', 'en')],
 ['cancer or benign tumor'],
 ['leukemia'],
 ['human disease'],
 ['disease'],
 ['hematopoietic and lymphoid system neoplasm'],
 [locstr('entity', 'en')],
 ['neoplastic disease or syndrome'],
 [locstr('realizable entity', 'en')],
 [locstr('continuant', 'en')],
 ['neoplasm'],
 [locstr('disposition', 'en')],
 [],
 ['hematologic disorder']]</code></pre>
</div>
</div>
</section>
<section id="get-child-concept-of-an-entity" class="level3">
<h3 class="anchored" data-anchor-id="get-child-concept-of-an-entity">Get “child” concept of an entity</h3>
<div id="cell-20" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:21:11.256398Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:21:11.256015Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:21:11.299972Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:21:11.298961Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:21:11.256364Z&quot;}" data-trusted="true" data-execution_count="8">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a>[x.label <span class="cf">for</span> x <span class="kw">in</span> entity.descendants()][<span class="dv">0</span>:<span class="dv">10</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="8">
<pre><code>[['chronic erythremia'],
 ['acute myelomonocytic leukemia M4'],
 ['acute myeloid leukemia, t(10;11)(p11.2;q23)'],
 ['hairy cell leukemia'],
 ['leukemia'],
 ['mixed phenotype acute leukemia with MLL rearranged'],
 ['acute myeloid leukemia, t(9;11)(p21.3;q23.3)'],
 ['acute myeloid leukemia, KIT exon 17 mutation'],
 ['therapy related acute myeloid leukemia and myelodysplastic syndrome'],
 ['acute myeloid leukemia, loss of chromosome 17p']]</code></pre>
</div>
</div>
<hr>
</section>
</section>
</section>
<section id="creating-sentencestore" class="level1">
<h1>Creating <strong>SentenceSTORE</strong></h1>
<p>Note that my goal is to recreate the knowledge stored in this ontology, by writing out the concepts above in plain English. I know this is what LLMs understand, so given the generous token length provided, I figured this is possible.</p>
<ul>
<li>“Cardiomyopathy is a phenotypic abnormality.”</li>
<li>“Histiocytoid cardiomyopathy is a cardiomyopathy.”</li>
</ul>
<p>These functions below basically do this task, grouping concepts for dealing with token length limits.</p>
<p>GOAL: - Take file path of the OWL files and optional predicate string override. - Turn <strong>[A]-(descendant)-[B]</strong> into <strong>“A is a B”</strong> - But allow Owlready2’s method names to be replaced with something that LLMs would recognize - My hunch was that “<strong>A descendant B</strong>” is not as definitive as “<strong>A is a B</strong>” for the sake of LLMs - it infers grammar / logic pretty well.</p>
<div id="cell-24" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:21:11.301620Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:21:11.301306Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:21:11.315882Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:21:11.314684Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:21:11.301590Z&quot;}" data-trusted="true" data-execution_count="9">
<div class="sourceCode cell-code" id="cb15"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb15-1"><a href="#cb15-1" aria-hidden="true" tabindex="-1"></a><span class="co"># This code itself was iteratively generated via LLM chabot, so bear with the ugliness</span></span>
<span id="cb15-2"><a href="#cb15-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_dbxref_relationships(entity):</span>
<span id="cb15-3"><a href="#cb15-3" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb15-4"><a href="#cb15-4" aria-hidden="true" tabindex="-1"></a><span class="co">    Get direct DbXref relationships for an entity.</span></span>
<span id="cb15-5"><a href="#cb15-5" aria-hidden="true" tabindex="-1"></a><span class="co">    </span></span>
<span id="cb15-6"><a href="#cb15-6" aria-hidden="true" tabindex="-1"></a><span class="co">    Args:</span></span>
<span id="cb15-7"><a href="#cb15-7" aria-hidden="true" tabindex="-1"></a><span class="co">        entity: The ontology entity to process</span></span>
<span id="cb15-8"><a href="#cb15-8" aria-hidden="true" tabindex="-1"></a><span class="co">        </span></span>
<span id="cb15-9"><a href="#cb15-9" aria-hidden="true" tabindex="-1"></a><span class="co">    Returns:</span></span>
<span id="cb15-10"><a href="#cb15-10" aria-hidden="true" tabindex="-1"></a><span class="co">        list: List of DbXref relationships</span></span>
<span id="cb15-11"><a href="#cb15-11" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb15-12"><a href="#cb15-12" aria-hidden="true" tabindex="-1"></a>    relationships <span class="op">=</span> []</span>
<span id="cb15-13"><a href="#cb15-13" aria-hidden="true" tabindex="-1"></a>    entity_label <span class="op">=</span> entity.label[<span class="dv">0</span>] <span class="cf">if</span> <span class="bu">hasattr</span>(entity, <span class="st">'label'</span>) <span class="kw">and</span> entity.label <span class="cf">else</span> entity.name</span>
<span id="cb15-14"><a href="#cb15-14" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb15-15"><a href="#cb15-15" aria-hidden="true" tabindex="-1"></a>    <span class="co"># </span><span class="al">TODO</span><span class="co"> This is ugly but was trial-and-error, someone rewrite it.</span></span>
<span id="cb15-16"><a href="#cb15-16" aria-hidden="true" tabindex="-1"></a>    <span class="co"># For the time being just wanted to focus on UMLS, SNOMECT, ICD10CM</span></span>
<span id="cb15-17"><a href="#cb15-17" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="bu">hasattr</span>(entity, <span class="st">'hasDbXref'</span>) <span class="kw">and</span> entity.hasDbXref:</span>
<span id="cb15-18"><a href="#cb15-18" aria-hidden="true" tabindex="-1"></a>        codes_as_string <span class="op">=</span> <span class="st">", "</span>.join([</span>
<span id="cb15-19"><a href="#cb15-19" aria-hidden="true" tabindex="-1"></a>            code.replace(<span class="st">":"</span>, <span class="st">' '</span>)</span>
<span id="cb15-20"><a href="#cb15-20" aria-hidden="true" tabindex="-1"></a>            .replace(<span class="st">'_2023_03_01'</span>, <span class="st">''</span>)     <span class="co"># Some SNOMED codes have this suffix</span></span>
<span id="cb15-21"><a href="#cb15-21" aria-hidden="true" tabindex="-1"></a>            .replace(<span class="st">'_CUI'</span>, <span class="st">''</span>)            <span class="co"># UMLS is prefixed as UMLS_CUI, feel unnecessary</span></span>
<span id="cb15-22"><a href="#cb15-22" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> code <span class="kw">in</span> entity.hasDbXref </span>
<span id="cb15-23"><a href="#cb15-23" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> code <span class="kw">and</span> <span class="bu">isinstance</span>(code, <span class="bu">str</span>) </span>
<span id="cb15-24"><a href="#cb15-24" aria-hidden="true" tabindex="-1"></a>                <span class="kw">and</span> <span class="bu">any</span>(code.startswith(prefix) <span class="cf">for</span> prefix <span class="kw">in</span> [<span class="st">'UMLS'</span>, <span class="st">'SNOMEDCT'</span>, <span class="st">'ICD10CM'</span>])</span>
<span id="cb15-25"><a href="#cb15-25" aria-hidden="true" tabindex="-1"></a>        ])</span>
<span id="cb15-26"><a href="#cb15-26" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> codes_as_string<span class="op">!=</span><span class="st">""</span>:</span>
<span id="cb15-27"><a href="#cb15-27" aria-hidden="true" tabindex="-1"></a>            relationships.append(<span class="ss">f"</span><span class="sc">{</span>entity_label<span class="sc">.</span>lower()<span class="sc">}</span><span class="ss"> can be coded with </span><span class="sc">{</span>codes_as_string<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb15-28"><a href="#cb15-28" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb15-29"><a href="#cb15-29" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> relationships</span>
<span id="cb15-30"><a href="#cb15-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-31"><a href="#cb15-31" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_entity_relationships(entity, relationship_map):</span>
<span id="cb15-32"><a href="#cb15-32" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb15-33"><a href="#cb15-33" aria-hidden="true" tabindex="-1"></a><span class="co">    Get unique direct relationships for an entity based on defined relationship mappings.</span></span>
<span id="cb15-34"><a href="#cb15-34" aria-hidden="true" tabindex="-1"></a><span class="co">    Excludes self-referential relationships.</span></span>
<span id="cb15-35"><a href="#cb15-35" aria-hidden="true" tabindex="-1"></a><span class="co">    </span></span>
<span id="cb15-36"><a href="#cb15-36" aria-hidden="true" tabindex="-1"></a><span class="co">    Args:</span></span>
<span id="cb15-37"><a href="#cb15-37" aria-hidden="true" tabindex="-1"></a><span class="co">        entity: The ontology entity to process</span></span>
<span id="cb15-38"><a href="#cb15-38" aria-hidden="true" tabindex="-1"></a><span class="co">        relationship_map: Dict mapping method names to relationship phrases</span></span>
<span id="cb15-39"><a href="#cb15-39" aria-hidden="true" tabindex="-1"></a><span class="co">            e.g. {"descendants": "is a", "parts": "is part of"}</span></span>
<span id="cb15-40"><a href="#cb15-40" aria-hidden="true" tabindex="-1"></a><span class="co">        </span></span>
<span id="cb15-41"><a href="#cb15-41" aria-hidden="true" tabindex="-1"></a><span class="co">    Returns:</span></span>
<span id="cb15-42"><a href="#cb15-42" aria-hidden="true" tabindex="-1"></a><span class="co">        list: List of unique direct relationships</span></span>
<span id="cb15-43"><a href="#cb15-43" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb15-44"><a href="#cb15-44" aria-hidden="true" tabindex="-1"></a>    relationships <span class="op">=</span> <span class="bu">set</span>()</span>
<span id="cb15-45"><a href="#cb15-45" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb15-46"><a href="#cb15-46" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Handle different label types including locstr</span></span>
<span id="cb15-47"><a href="#cb15-47" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> get_label(e):</span>
<span id="cb15-48"><a href="#cb15-48" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="bu">hasattr</span>(e, <span class="st">'label'</span>) <span class="kw">and</span> e.label:</span>
<span id="cb15-49"><a href="#cb15-49" aria-hidden="true" tabindex="-1"></a>            <span class="co"># locstr objects can be used directly as strings</span></span>
<span id="cb15-50"><a href="#cb15-50" aria-hidden="true" tabindex="-1"></a>            <span class="cf">return</span> <span class="bu">str</span>(e.label[<span class="dv">0</span>])</span>
<span id="cb15-51"><a href="#cb15-51" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> e.name</span>
<span id="cb15-52"><a href="#cb15-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-53"><a href="#cb15-53" aria-hidden="true" tabindex="-1"></a>    entity_label <span class="op">=</span> get_label(entity)</span>
<span id="cb15-54"><a href="#cb15-54" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-55"><a href="#cb15-55" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Iterate through each relationship type defined in the mapping</span></span>
<span id="cb15-56"><a href="#cb15-56" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> method_name, relationship_phrase <span class="kw">in</span> relationship_map.items():</span>
<span id="cb15-57"><a href="#cb15-57" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Get the method if it exists on the entity</span></span>
<span id="cb15-58"><a href="#cb15-58" aria-hidden="true" tabindex="-1"></a>        relation_method <span class="op">=</span> <span class="bu">getattr</span>(entity, method_name, <span class="va">None</span>)</span>
<span id="cb15-59"><a href="#cb15-59" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb15-60"><a href="#cb15-60" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> relation_method <span class="kw">and</span> <span class="bu">callable</span>(relation_method):</span>
<span id="cb15-61"><a href="#cb15-61" aria-hidden="true" tabindex="-1"></a>            <span class="co"># Get related entities using the method</span></span>
<span id="cb15-62"><a href="#cb15-62" aria-hidden="true" tabindex="-1"></a>            related_entities <span class="op">=</span> relation_method()</span>
<span id="cb15-63"><a href="#cb15-63" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb15-64"><a href="#cb15-64" aria-hidden="true" tabindex="-1"></a>            <span class="co"># Handle both iterator and list returns</span></span>
<span id="cb15-65"><a href="#cb15-65" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> <span class="kw">not</span> <span class="bu">isinstance</span>(related_entities, (<span class="bu">list</span>, <span class="bu">set</span>)):</span>
<span id="cb15-66"><a href="#cb15-66" aria-hidden="true" tabindex="-1"></a>                related_entities <span class="op">=</span> <span class="bu">list</span>(related_entities)</span>
<span id="cb15-67"><a href="#cb15-67" aria-hidden="true" tabindex="-1"></a>                </span>
<span id="cb15-68"><a href="#cb15-68" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> related <span class="kw">in</span> related_entities:</span>
<span id="cb15-69"><a href="#cb15-69" aria-hidden="true" tabindex="-1"></a>                rel_label <span class="op">=</span> get_label(related)</span>
<span id="cb15-70"><a href="#cb15-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-71"><a href="#cb15-71" aria-hidden="true" tabindex="-1"></a>                <span class="co"># Skip if the labels are identical (self-reference)</span></span>
<span id="cb15-72"><a href="#cb15-72" aria-hidden="true" tabindex="-1"></a>                <span class="co"># Don't want "Diabetes is diabetes"</span></span>
<span id="cb15-73"><a href="#cb15-73" aria-hidden="true" tabindex="-1"></a>                <span class="cf">if</span> rel_label <span class="op">!=</span> entity_label:</span>
<span id="cb15-74"><a href="#cb15-74" aria-hidden="true" tabindex="-1"></a>                    relationships.add(<span class="ss">f"</span><span class="sc">{</span>rel_label<span class="sc">.</span>lower()<span class="sc">}</span><span class="ss"> </span><span class="sc">{</span>relationship_phrase<span class="sc">}</span><span class="ss"> </span><span class="sc">{</span>entity_label<span class="sc">.</span>lower()<span class="sc">}</span><span class="ss">"</span>)</span>
<span id="cb15-75"><a href="#cb15-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-76"><a href="#cb15-76" aria-hidden="true" tabindex="-1"></a>    dbxrefs <span class="op">=</span> []</span>
<span id="cb15-77"><a href="#cb15-77" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-78"><a href="#cb15-78" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Get DbXref relationships (for entity and all descendants)</span></span>
<span id="cb15-79"><a href="#cb15-79" aria-hidden="true" tabindex="-1"></a>    visited <span class="op">=</span> <span class="bu">set</span>()</span>
<span id="cb15-80"><a href="#cb15-80" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> process_dbxrefs(current_entity):</span>
<span id="cb15-81"><a href="#cb15-81" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> current_entity <span class="kw">not</span> <span class="kw">in</span> visited:</span>
<span id="cb15-82"><a href="#cb15-82" aria-hidden="true" tabindex="-1"></a>            visited.add(current_entity)</span>
<span id="cb15-83"><a href="#cb15-83" aria-hidden="true" tabindex="-1"></a>            dbxrefs.extend(get_dbxref_relationships(current_entity))</span>
<span id="cb15-84"><a href="#cb15-84" aria-hidden="true" tabindex="-1"></a>            <span class="cf">for</span> descendant <span class="kw">in</span> current_entity.descendants():</span>
<span id="cb15-85"><a href="#cb15-85" aria-hidden="true" tabindex="-1"></a>                process_dbxrefs(descendant)</span>
<span id="cb15-86"><a href="#cb15-86" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> dbxrefs</span>
<span id="cb15-87"><a href="#cb15-87" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb15-88"><a href="#cb15-88" aria-hidden="true" tabindex="-1"></a>    cross_reference_relationships <span class="op">=</span> <span class="bu">set</span>(process_dbxrefs(entity))</span>
<span id="cb15-89"><a href="#cb15-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-90"><a href="#cb15-90" aria-hidden="true" tabindex="-1"></a>    relationships <span class="op">=</span> relationships.union(cross_reference_relationships)</span>
<span id="cb15-91"><a href="#cb15-91" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb15-92"><a href="#cb15-92" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> <span class="bu">list</span>(relationships)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
<section id="test-cross-reference" class="level2">
<h2 class="anchored" data-anchor-id="test-cross-reference">Test cross-reference</h2>
<div id="cell-27" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:21:11.318020Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:21:11.317669Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:21:11.451044Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:21:11.450062Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:21:11.317990Z&quot;}" data-trusted="true" data-execution_count="10">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a>search_result <span class="op">=</span> onto.search(label<span class="op">=</span><span class="st">"*iabete*"</span>)</span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a>entity <span class="op">=</span> search_result[<span class="dv">0</span>]</span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a>entity</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="10">
<pre><code>obo.HP_0000819</code></pre>
</div>
</div>
<div id="cell-28" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:21:11.454249Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:21:11.453832Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:21:11.460277Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:21:11.459322Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:21:11.454217Z&quot;}" data-trusted="true" data-execution_count="11">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a>entity.label</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="11">
<pre><code>['Diabetes mellitus', 'Diabetes mellitus']</code></pre>
</div>
</div>
<div id="cell-29" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:21:11.461946Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:21:11.461534Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:21:11.470919Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:21:11.469976Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:21:11.461903Z&quot;}" data-trusted="true" data-execution_count="12">
<div class="sourceCode cell-code" id="cb20"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb20-1"><a href="#cb20-1" aria-hidden="true" tabindex="-1"></a>entity.hasDbXref</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="12">
<pre><code>['SNOMEDCT_US:73211009', 'UMLS:C0011849']</code></pre>
</div>
</div>
</section>
<section id="test-define-the-relationship" class="level2">
<h2 class="anchored" data-anchor-id="test-define-the-relationship">Test “Define the relationship!”</h2>
<p>Have to make sure the entities get written out as english language sentence.</p>
<div id="cell-31" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:22:05.401915Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:22:05.400983Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:22:05.410070Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:22:05.408912Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:22:05.401862Z&quot;}" data-trusted="true" data-execution_count="13">
<div class="sourceCode cell-code" id="cb22"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb22-1"><a href="#cb22-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Key: method name for owlready2</span></span>
<span id="cb22-2"><a href="#cb22-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Value: what string to use for defining the relationship (DTR) </span></span>
<span id="cb22-3"><a href="#cb22-3" aria-hidden="true" tabindex="-1"></a>RELATIONSHIP_MAP <span class="op">=</span> {</span>
<span id="cb22-4"><a href="#cb22-4" aria-hidden="true" tabindex="-1"></a>    <span class="st">"subclasses"</span>: <span class="st">"is a"</span>,</span>
<span id="cb22-5"><a href="#cb22-5" aria-hidden="true" tabindex="-1"></a>    <span class="st">"parts"</span>: <span class="st">"is part of"</span>,</span>
<span id="cb22-6"><a href="#cb22-6" aria-hidden="true" tabindex="-1"></a>    <span class="st">"connected_to"</span>: <span class="st">"connects to"</span>,</span>
<span id="cb22-7"><a href="#cb22-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">"located_in"</span>: <span class="st">"is located in"</span></span>
<span id="cb22-8"><a href="#cb22-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb22-9"><a href="#cb22-9" aria-hidden="true" tabindex="-1"></a>get_entity_relationships(entity, relationship_map<span class="op">=</span>RELATIONSHIP_MAP)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="13">
<pre><code>['neonatal insulin-dependent diabetes mellitus can be coded with UMLS C3278636',
 'maturity-onset diabetes of the young can be coded with SNOMEDCT_US 609561005, UMLS C0342276',
 'type i diabetes mellitus is a diabetes mellitus',
 'insulin-dependent but ketosis-resistant diabetes can be coded with UMLS C1842404',
 'type i diabetes mellitus can be coded with SNOMEDCT_US 46635009, UMLS C0011854',
 'diabetic ketoacidosis can be coded with SNOMEDCT_US 420422005, UMLS C0011880',
 'insulin-resistant diabetes mellitus at puberty can be coded with UMLS C1837792',
 'transient neonatal diabetes mellitus can be coded with SNOMEDCT_US 237603002, UMLS C0342273',
 'type ii diabetes mellitus can be coded with SNOMEDCT_US 44054006, UMLS C0011860',
 'maturity-onset diabetes of the young is a diabetes mellitus',
 'diabetes mellitus can be coded with SNOMEDCT_US 73211009, UMLS C0011849',
 'insulin-resistant diabetes mellitus can be coded with UMLS C0854110',
 'insulin-resistant diabetes mellitus is a diabetes mellitus',
 'diabetic ketoacidosis is a diabetes mellitus',
 'type ii diabetes mellitus is a diabetes mellitus']</code></pre>
</div>
</div>
</section>
<section id="sentencestore-is-ready" class="level2">
<h2 class="anchored" data-anchor-id="sentencestore-is-ready">SentenceSTORE is READY!</h2>
<div class="img-fluid quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="ai-ontology-entity-linking-for-clinical-content_files/figure-html/cell-33-1-24e64324-33a7-4730-a9a4-c3049de79cbe.png" class="img-fluid figure-img"></p>
<figcaption>image.png</figcaption>
</figure>
</div>
<hr>
</section>
</section>
<section id="systematically-process-the-ontologies" class="level1">
<h1>Systematically process the ontologies</h1>
<p>Since the single entity extracton work, we need to use the processing function to extract the whole thing.</p>
<p>For each ontology there are a handful of “top-level” entities that need to be selected. For example:</p>
<div class="img-fluid quarto-figure quarto-figure-center">
<figure class="figure">
<p><img src="ai-ontology-entity-linking-for-clinical-content_files/figure-html/cell-36-1-3cb0db72-2c47-44d1-8395-6d063d60960f.png" class="img-fluid figure-img"></p>
<figcaption>image.png</figcaption>
</figure>
</div>
<p>Since all ontologies have a hierachical structure, using the top level concept, traversing it down literatively with children concepts, we can cover all the ontology entities.</p>
<section id="ontology-.owl-to-sentencestore-converter" class="level2">
<h2 class="anchored" data-anchor-id="ontology-.owl-to-sentencestore-converter">Ontology (.owl) to SentenceSTORE Converter</h2>
<div id="cell-39" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:22:06.777411Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:22:06.776686Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:22:06.787312Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:22:06.786381Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:22:06.777372Z&quot;}" data-trusted="true" data-execution_count="14">
<div class="sourceCode cell-code" id="cb24"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb24-1"><a href="#cb24-1" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> process_ontology_relationships(ontology_path, top_level_labels<span class="op">=</span><span class="va">None</span>, relationship_map<span class="op">=</span><span class="va">None</span>):</span>
<span id="cb24-2"><a href="#cb24-2" aria-hidden="true" tabindex="-1"></a>    <span class="co">"""</span></span>
<span id="cb24-3"><a href="#cb24-3" aria-hidden="true" tabindex="-1"></a><span class="co">    Process ontology relationships recursively starting from specified top-level entities.</span></span>
<span id="cb24-4"><a href="#cb24-4" aria-hidden="true" tabindex="-1"></a><span class="co">    Returns sorted unique relationships across all specified entities and their subclasses.</span></span>
<span id="cb24-5"><a href="#cb24-5" aria-hidden="true" tabindex="-1"></a><span class="co">    </span></span>
<span id="cb24-6"><a href="#cb24-6" aria-hidden="true" tabindex="-1"></a><span class="co">    Args:</span></span>
<span id="cb24-7"><a href="#cb24-7" aria-hidden="true" tabindex="-1"></a><span class="co">        ontology_path: str, Path to the ontology file</span></span>
<span id="cb24-8"><a href="#cb24-8" aria-hidden="true" tabindex="-1"></a><span class="co">        top_level_labels: list[str], Labels of top-level entries to process.</span></span>
<span id="cb24-9"><a href="#cb24-9" aria-hidden="true" tabindex="-1"></a><span class="co">            If None, an empty list will be returned</span></span>
<span id="cb24-10"><a href="#cb24-10" aria-hidden="true" tabindex="-1"></a><span class="co">        relationship_map: Optional[dict], Mapping of method names to relationship phrases</span></span>
<span id="cb24-11"><a href="#cb24-11" aria-hidden="true" tabindex="-1"></a><span class="co">            </span></span>
<span id="cb24-12"><a href="#cb24-12" aria-hidden="true" tabindex="-1"></a><span class="co">    Returns:</span></span>
<span id="cb24-13"><a href="#cb24-13" aria-hidden="true" tabindex="-1"></a><span class="co">        list: Sorted unique relationships found across specified top-level entities and their subclasses</span></span>
<span id="cb24-14"><a href="#cb24-14" aria-hidden="true" tabindex="-1"></a><span class="co">    """</span></span>
<span id="cb24-15"><a href="#cb24-15" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Default relationship mapping</span></span>
<span id="cb24-16"><a href="#cb24-16" aria-hidden="true" tabindex="-1"></a>    DEFAULT_RELATIONSHIP_MAP <span class="op">=</span> {</span>
<span id="cb24-17"><a href="#cb24-17" aria-hidden="true" tabindex="-1"></a>        <span class="st">"subclasses"</span>: <span class="st">"is a"</span>,</span>
<span id="cb24-18"><a href="#cb24-18" aria-hidden="true" tabindex="-1"></a>        <span class="st">"parts"</span>: <span class="st">"is part of"</span>,</span>
<span id="cb24-19"><a href="#cb24-19" aria-hidden="true" tabindex="-1"></a>        <span class="st">"connected_to"</span>: <span class="st">"connects to"</span>,</span>
<span id="cb24-20"><a href="#cb24-20" aria-hidden="true" tabindex="-1"></a>        <span class="st">"located_in"</span>: <span class="st">"is located in"</span></span>
<span id="cb24-21"><a href="#cb24-21" aria-hidden="true" tabindex="-1"></a>    }</span>
<span id="cb24-22"><a href="#cb24-22" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb24-23"><a href="#cb24-23" aria-hidden="true" tabindex="-1"></a>    <span class="im">from</span> owlready2 <span class="im">import</span> get_ontology, get_namespace</span>
<span id="cb24-24"><a href="#cb24-24" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb24-25"><a href="#cb24-25" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Load ontology</span></span>
<span id="cb24-26"><a href="#cb24-26" aria-hidden="true" tabindex="-1"></a>    onto <span class="op">=</span> get_ontology(ontology_path).load()</span>
<span id="cb24-27"><a href="#cb24-27" aria-hidden="true" tabindex="-1"></a>    obo <span class="op">=</span> get_namespace(<span class="st">"http://purl.obolibrary.org/obo/"</span>)</span>
<span id="cb24-28"><a href="#cb24-28" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb24-29"><a href="#cb24-29" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Use provided or default relationship map</span></span>
<span id="cb24-30"><a href="#cb24-30" aria-hidden="true" tabindex="-1"></a>    relationship_map <span class="op">=</span> relationship_map <span class="kw">or</span> DEFAULT_RELATIONSHIP_MAP</span>
<span id="cb24-31"><a href="#cb24-31" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb24-32"><a href="#cb24-32" aria-hidden="true" tabindex="-1"></a>    <span class="cf">if</span> <span class="kw">not</span> top_level_labels:</span>
<span id="cb24-33"><a href="#cb24-33" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> []</span>
<span id="cb24-34"><a href="#cb24-34" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb24-35"><a href="#cb24-35" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">"</span><span class="ch">\n</span><span class="st">Processing relationships:"</span>)</span>
<span id="cb24-36"><a href="#cb24-36" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">"-"</span> <span class="op">*</span> <span class="dv">40</span>)</span>
<span id="cb24-37"><a href="#cb24-37" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb24-38"><a href="#cb24-38" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> process_entity_and_subclasses(entity, processed_entities<span class="op">=</span><span class="va">None</span>):</span>
<span id="cb24-39"><a href="#cb24-39" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""</span></span>
<span id="cb24-40"><a href="#cb24-40" aria-hidden="true" tabindex="-1"></a><span class="co">        Recursively process an entity and all its subclasses.</span></span>
<span id="cb24-41"><a href="#cb24-41" aria-hidden="true" tabindex="-1"></a><span class="co">        </span></span>
<span id="cb24-42"><a href="#cb24-42" aria-hidden="true" tabindex="-1"></a><span class="co">        Args:</span></span>
<span id="cb24-43"><a href="#cb24-43" aria-hidden="true" tabindex="-1"></a><span class="co">            entity: The entity to process</span></span>
<span id="cb24-44"><a href="#cb24-44" aria-hidden="true" tabindex="-1"></a><span class="co">            processed_entities: Set of already processed entities to avoid cycles</span></span>
<span id="cb24-45"><a href="#cb24-45" aria-hidden="true" tabindex="-1"></a><span class="co">            </span></span>
<span id="cb24-46"><a href="#cb24-46" aria-hidden="true" tabindex="-1"></a><span class="co">        Returns:</span></span>
<span id="cb24-47"><a href="#cb24-47" aria-hidden="true" tabindex="-1"></a><span class="co">            list: All relationships found for this entity and its subclasses</span></span>
<span id="cb24-48"><a href="#cb24-48" aria-hidden="true" tabindex="-1"></a><span class="co">        """</span></span>
<span id="cb24-49"><a href="#cb24-49" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> processed_entities <span class="kw">is</span> <span class="va">None</span>:</span>
<span id="cb24-50"><a href="#cb24-50" aria-hidden="true" tabindex="-1"></a>            processed_entities <span class="op">=</span> <span class="bu">set</span>()</span>
<span id="cb24-51"><a href="#cb24-51" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb24-52"><a href="#cb24-52" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> entity <span class="kw">in</span> processed_entities:</span>
<span id="cb24-53"><a href="#cb24-53" aria-hidden="true" tabindex="-1"></a>            <span class="cf">return</span> []</span>
<span id="cb24-54"><a href="#cb24-54" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb24-55"><a href="#cb24-55" aria-hidden="true" tabindex="-1"></a>        processed_entities.add(entity)</span>
<span id="cb24-56"><a href="#cb24-56" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb24-57"><a href="#cb24-57" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Get relationships for current entity</span></span>
<span id="cb24-58"><a href="#cb24-58" aria-hidden="true" tabindex="-1"></a>        relationships <span class="op">=</span> get_entity_relationships(entity, relationship_map)</span>
<span id="cb24-59"><a href="#cb24-59" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb24-60"><a href="#cb24-60" aria-hidden="true" tabindex="-1"></a>        <span class="co"># Process all subclasses</span></span>
<span id="cb24-61"><a href="#cb24-61" aria-hidden="true" tabindex="-1"></a>        <span class="cf">for</span> subclass <span class="kw">in</span> entity.descendants():</span>
<span id="cb24-62"><a href="#cb24-62" aria-hidden="true" tabindex="-1"></a>            <span class="cf">if</span> subclass <span class="kw">not</span> <span class="kw">in</span> processed_entities:</span>
<span id="cb24-63"><a href="#cb24-63" aria-hidden="true" tabindex="-1"></a>                sub_relationships <span class="op">=</span> process_entity_and_subclasses(subclass, processed_entities)</span>
<span id="cb24-64"><a href="#cb24-64" aria-hidden="true" tabindex="-1"></a>                relationships.extend(sub_relationships)</span>
<span id="cb24-65"><a href="#cb24-65" aria-hidden="true" tabindex="-1"></a>                </span>
<span id="cb24-66"><a href="#cb24-66" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> relationships</span>
<span id="cb24-67"><a href="#cb24-67" aria-hidden="true" tabindex="-1"></a>        </span>
<span id="cb24-68"><a href="#cb24-68" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Find and process top-level entities</span></span>
<span id="cb24-69"><a href="#cb24-69" aria-hidden="true" tabindex="-1"></a>    all_relationships <span class="op">=</span> []</span>
<span id="cb24-70"><a href="#cb24-70" aria-hidden="true" tabindex="-1"></a>    processed_entities <span class="op">=</span> <span class="bu">set</span>()</span>
<span id="cb24-71"><a href="#cb24-71" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb24-72"><a href="#cb24-72" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> label <span class="kw">in</span> top_level_labels:</span>
<span id="cb24-73"><a href="#cb24-73" aria-hidden="true" tabindex="-1"></a>        search_results <span class="op">=</span> onto.search(label<span class="op">=</span>label)</span>
<span id="cb24-74"><a href="#cb24-74" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> search_results:</span>
<span id="cb24-75"><a href="#cb24-75" aria-hidden="true" tabindex="-1"></a>            top_entity <span class="op">=</span> search_results[<span class="dv">0</span>]</span>
<span id="cb24-76"><a href="#cb24-76" aria-hidden="true" tabindex="-1"></a>            relationships <span class="op">=</span> process_entity_and_subclasses(top_entity, processed_entities)</span>
<span id="cb24-77"><a href="#cb24-77" aria-hidden="true" tabindex="-1"></a>            all_relationships.extend(relationships)</span>
<span id="cb24-78"><a href="#cb24-78" aria-hidden="true" tabindex="-1"></a>            <span class="bu">print</span>(<span class="ss">f"</span><span class="sc">{</span>label<span class="sc">}</span><span class="ss">: </span><span class="sc">{</span><span class="bu">len</span>(relationships)<span class="sc">}</span><span class="ss"> relationships found"</span>)</span>
<span id="cb24-79"><a href="#cb24-79" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-80"><a href="#cb24-80" aria-hidden="true" tabindex="-1"></a>    <span class="co"># Remove duplicates and sort</span></span>
<span id="cb24-81"><a href="#cb24-81" aria-hidden="true" tabindex="-1"></a>    unique_relationships <span class="op">=</span> <span class="bu">sorted</span>(<span class="bu">list</span>(<span class="bu">set</span>(all_relationships)))</span>
<span id="cb24-82"><a href="#cb24-82" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-83"><a href="#cb24-83" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">"-"</span> <span class="op">*</span> <span class="dv">40</span>)</span>
<span id="cb24-84"><a href="#cb24-84" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f"Total unique relationships: </span><span class="sc">{</span><span class="bu">len</span>(unique_relationships)<span class="sc">}</span><span class="ch">\n</span><span class="ss">"</span>)</span>
<span id="cb24-85"><a href="#cb24-85" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb24-86"><a href="#cb24-86" aria-hidden="true" tabindex="-1"></a>    <span class="cf">return</span> unique_relationships</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="hpo" class="level2">
<h2 class="anchored" data-anchor-id="hpo">1. HPO</h2>
<p>I iteratively cherry picked these for performance, as Kaggle gives me trouble. Personal notebook works.</p>
<div id="cell-41" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:22:07.124474Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:22:07.124073Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:22:29.198390Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:22:29.197124Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:22:07.124439Z&quot;}" data-trusted="true" data-execution_count="15">
<div class="sourceCode cell-code" id="cb25"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb25-1"><a href="#cb25-1" aria-hidden="true" tabindex="-1"></a>relationships_from_hpo <span class="op">=</span> process_ontology_relationships(</span>
<span id="cb25-2"><a href="#cb25-2" aria-hidden="true" tabindex="-1"></a>    ontology_path<span class="op">=</span><span class="st">"ontologies/hp.owl"</span>,</span>
<span id="cb25-3"><a href="#cb25-3" aria-hidden="true" tabindex="-1"></a>    top_level_labels<span class="op">=</span>[</span>
<span id="cb25-4"><a href="#cb25-4" aria-hidden="true" tabindex="-1"></a>        <span class="co"># "Biospecimen phenotypic feature",</span></span>
<span id="cb25-5"><a href="#cb25-5" aria-hidden="true" tabindex="-1"></a>        <span class="co"># "Blood group",   </span></span>
<span id="cb25-6"><a href="#cb25-6" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Clinical modifier"</span>,</span>
<span id="cb25-7"><a href="#cb25-7" aria-hidden="true" tabindex="-1"></a>        <span class="co"># "Mode of inheritance",</span></span>
<span id="cb25-8"><a href="#cb25-8" aria-hidden="true" tabindex="-1"></a>        <span class="co"># "Frequency",</span></span>
<span id="cb25-9"><a href="#cb25-9" aria-hidden="true" tabindex="-1"></a>        <span class="co"># "Past medical history",</span></span>
<span id="cb25-10"><a href="#cb25-10" aria-hidden="true" tabindex="-1"></a>        <span class="st">"Phenotypic abnormality"</span></span>
<span id="cb25-11"><a href="#cb25-11" aria-hidden="true" tabindex="-1"></a>    ],</span>
<span id="cb25-12"><a href="#cb25-12" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Processing relationships:
----------------------------------------
Clinical modifier: 571 relationships found
Phenotypic abnormality: 155221 relationships found
----------------------------------------
Total unique relationships: 34581
</code></pre>
</div>
</div>
</section>
<section id="mondo" class="level2">
<h2 class="anchored" data-anchor-id="mondo">2. MONDO</h2>
<div id="cell-43" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:22:29.200514Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:22:29.200193Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:22:50.823250Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:22:50.822171Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:22:29.200484Z&quot;}" data-trusted="true" data-execution_count="16">
<div class="sourceCode cell-code" id="cb27"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb27-1"><a href="#cb27-1" aria-hidden="true" tabindex="-1"></a>relationships_from_mondo <span class="op">=</span> process_ontology_relationships(</span>
<span id="cb27-2"><a href="#cb27-2" aria-hidden="true" tabindex="-1"></a>    ontology_path<span class="op">=</span><span class="st">"ontologies/mondo.owl"</span>,</span>
<span id="cb27-3"><a href="#cb27-3" aria-hidden="true" tabindex="-1"></a>    top_level_labels<span class="op">=</span>[</span>
<span id="cb27-4"><a href="#cb27-4" aria-hidden="true" tabindex="-1"></a>        <span class="co"># "disease",</span></span>
<span id="cb27-5"><a href="#cb27-5" aria-hidden="true" tabindex="-1"></a>        <span class="st">"human disease"</span>, </span>
<span id="cb27-6"><a href="#cb27-6" aria-hidden="true" tabindex="-1"></a>        <span class="co"># "disease characteristic",</span></span>
<span id="cb27-7"><a href="#cb27-7" aria-hidden="true" tabindex="-1"></a>        <span class="co"># "disease susceptibility",  # Has long names</span></span>
<span id="cb27-8"><a href="#cb27-8" aria-hidden="true" tabindex="-1"></a>        <span class="st">"injury"</span></span>
<span id="cb27-9"><a href="#cb27-9" aria-hidden="true" tabindex="-1"></a>    ],</span>
<span id="cb27-10"><a href="#cb27-10" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Processing relationships:
----------------------------------------
human disease: 281182 relationships found
injury: 23 relationships found
----------------------------------------
Total unique relationships: 55138
</code></pre>
</div>
</div>
</section>
<section id="disease-ontology" class="level2">
<h2 class="anchored" data-anchor-id="disease-ontology">3. Disease Ontology</h2>
<div id="cell-45" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:22:50.825032Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:22:50.824597Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:22:59.937446Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:22:59.936279Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:22:50.824988Z&quot;}" data-trusted="true" data-execution_count="17">
<div class="sourceCode cell-code" id="cb29"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb29-1"><a href="#cb29-1" aria-hidden="true" tabindex="-1"></a>relationships_from_do <span class="op">=</span> process_ontology_relationships(</span>
<span id="cb29-2"><a href="#cb29-2" aria-hidden="true" tabindex="-1"></a>    ontology_path<span class="op">=</span><span class="st">"ontologies/doid-merged.owl"</span>,</span>
<span id="cb29-3"><a href="#cb29-3" aria-hidden="true" tabindex="-1"></a>    top_level_labels<span class="op">=</span>[</span>
<span id="cb29-4"><a href="#cb29-4" aria-hidden="true" tabindex="-1"></a>        <span class="st">"disease by infectious agent"</span>,</span>
<span id="cb29-5"><a href="#cb29-5" aria-hidden="true" tabindex="-1"></a>        <span class="st">"disease of anatomical entity"</span>,</span>
<span id="cb29-6"><a href="#cb29-6" aria-hidden="true" tabindex="-1"></a>        <span class="st">"disease of cellular proliferation"</span>,</span>
<span id="cb29-7"><a href="#cb29-7" aria-hidden="true" tabindex="-1"></a>        <span class="co"># "disease of mental health",</span></span>
<span id="cb29-8"><a href="#cb29-8" aria-hidden="true" tabindex="-1"></a>        <span class="st">"disease of metabolism"</span>,</span>
<span id="cb29-9"><a href="#cb29-9" aria-hidden="true" tabindex="-1"></a>        <span class="co"># "genetic disease",</span></span>
<span id="cb29-10"><a href="#cb29-10" aria-hidden="true" tabindex="-1"></a>        <span class="st">"physical disorder"</span>,</span>
<span id="cb29-11"><a href="#cb29-11" aria-hidden="true" tabindex="-1"></a>        <span class="st">"syndrome"</span></span>
<span id="cb29-12"><a href="#cb29-12" aria-hidden="true" tabindex="-1"></a>    ],</span>
<span id="cb29-13"><a href="#cb29-13" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>
Processing relationships:
----------------------------------------
disease by infectious agent: 2272 relationships found
disease of anatomical entity: 53917 relationships found
disease of cellular proliferation: 10565 relationships found
disease of metabolism: 2902 relationships found
physical disorder: 382 relationships found
syndrome: 2023 relationships found
----------------------------------------
Total unique relationships: 19803
</code></pre>
</div>
</div>
<section id="option-utility-polars-search-function" class="level3">
<h3 class="anchored" data-anchor-id="option-utility-polars-search-function">[Option Utility]: Polars search function</h3>
<blockquote class="blockquote">
<p>Just to make string search easier</p>
</blockquote>
<div id="cell-47" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:22:59.939841Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:22:59.939490Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:23:00.339803Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:23:00.338792Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:22:59.939809Z&quot;}" data-trusted="true" data-execution_count="18">
<div class="sourceCode cell-code" id="cb31"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb31-1"><a href="#cb31-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> polars <span class="im">as</span> pl</span>
<span id="cb31-2"><a href="#cb31-2" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> typing <span class="im">import</span> Union, Pattern, List</span>
<span id="cb31-3"><a href="#cb31-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-4"><a href="#cb31-4" aria-hidden="true" tabindex="-1"></a><span class="co"># All ontology search</span></span>
<span id="cb31-5"><a href="#cb31-5" aria-hidden="true" tabindex="-1"></a><span class="co"># These are just for sanity checking proper processing.</span></span>
<span id="cb31-6"><a href="#cb31-6" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> pl.from_dict({</span>
<span id="cb31-7"><a href="#cb31-7" aria-hidden="true" tabindex="-1"></a>    <span class="st">"text_relationships"</span>: <span class="bu">list</span>(<span class="bu">set</span>(</span>
<span id="cb31-8"><a href="#cb31-8" aria-hidden="true" tabindex="-1"></a>        relationships_from_hpo <span class="op">+</span> </span>
<span id="cb31-9"><a href="#cb31-9" aria-hidden="true" tabindex="-1"></a>        relationships_from_do <span class="op">+</span> </span>
<span id="cb31-10"><a href="#cb31-10" aria-hidden="true" tabindex="-1"></a>        relationships_from_mondo</span>
<span id="cb31-11"><a href="#cb31-11" aria-hidden="true" tabindex="-1"></a>    ))})</span>
<span id="cb31-12"><a href="#cb31-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-13"><a href="#cb31-13" aria-hidden="true" tabindex="-1"></a>hpo_df <span class="op">=</span> pl.from_dict({</span>
<span id="cb31-14"><a href="#cb31-14" aria-hidden="true" tabindex="-1"></a>    <span class="st">"text_relationships"</span>: <span class="bu">list</span>(<span class="bu">set</span>(</span>
<span id="cb31-15"><a href="#cb31-15" aria-hidden="true" tabindex="-1"></a>        relationships_from_hpo</span>
<span id="cb31-16"><a href="#cb31-16" aria-hidden="true" tabindex="-1"></a>    ))})</span>
<span id="cb31-17"><a href="#cb31-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-18"><a href="#cb31-18" aria-hidden="true" tabindex="-1"></a>mondo_df <span class="op">=</span> pl.from_dict({</span>
<span id="cb31-19"><a href="#cb31-19" aria-hidden="true" tabindex="-1"></a>    <span class="st">"text_relationships"</span>: <span class="bu">list</span>(<span class="bu">set</span>(</span>
<span id="cb31-20"><a href="#cb31-20" aria-hidden="true" tabindex="-1"></a>        relationships_from_mondo</span>
<span id="cb31-21"><a href="#cb31-21" aria-hidden="true" tabindex="-1"></a>    ))})</span>
<span id="cb31-22"><a href="#cb31-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-23"><a href="#cb31-23" aria-hidden="true" tabindex="-1"></a>do_df <span class="op">=</span> pl.from_dict({</span>
<span id="cb31-24"><a href="#cb31-24" aria-hidden="true" tabindex="-1"></a>    <span class="st">"text_relationships"</span>: <span class="bu">list</span>(<span class="bu">set</span>(</span>
<span id="cb31-25"><a href="#cb31-25" aria-hidden="true" tabindex="-1"></a>        relationships_from_do</span>
<span id="cb31-26"><a href="#cb31-26" aria-hidden="true" tabindex="-1"></a>    ))})</span>
<span id="cb31-27"><a href="#cb31-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb31-28"><a href="#cb31-28" aria-hidden="true" tabindex="-1"></a><span class="at">@pl.api.register_dataframe_namespace</span>(<span class="st">"regex"</span>)</span>
<span id="cb31-29"><a href="#cb31-29" aria-hidden="true" tabindex="-1"></a><span class="kw">class</span> RegexFrame:</span>
<span id="cb31-30"><a href="#cb31-30" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> <span class="fu">__init__</span>(<span class="va">self</span>, df: pl.DataFrame) <span class="op">-&gt;</span> List:</span>
<span id="cb31-31"><a href="#cb31-31" aria-hidden="true" tabindex="-1"></a>        <span class="va">self</span>._df <span class="op">=</span> df</span>
<span id="cb31-32"><a href="#cb31-32" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb31-33"><a href="#cb31-33" aria-hidden="true" tabindex="-1"></a>    <span class="kw">def</span> search(<span class="va">self</span>, </span>
<span id="cb31-34"><a href="#cb31-34" aria-hidden="true" tabindex="-1"></a>              pattern: Union[<span class="bu">str</span>, Pattern],</span>
<span id="cb31-35"><a href="#cb31-35" aria-hidden="true" tabindex="-1"></a>              column: <span class="bu">str</span>) <span class="op">-&gt;</span> List[<span class="bu">str</span>]:</span>
<span id="cb31-36"><a href="#cb31-36" aria-hidden="true" tabindex="-1"></a>        <span class="co">"""Search column using regex pattern. Returns sorted list of matching strings."""</span></span>
<span id="cb31-37"><a href="#cb31-37" aria-hidden="true" tabindex="-1"></a>        <span class="cf">if</span> <span class="bu">isinstance</span>(pattern, <span class="bu">str</span>):</span>
<span id="cb31-38"><a href="#cb31-38" aria-hidden="true" tabindex="-1"></a>            search_pat <span class="op">=</span> pattern</span>
<span id="cb31-39"><a href="#cb31-39" aria-hidden="true" tabindex="-1"></a>        <span class="cf">else</span>:</span>
<span id="cb31-40"><a href="#cb31-40" aria-hidden="true" tabindex="-1"></a>            search_pat <span class="op">=</span> pattern.pattern</span>
<span id="cb31-41"><a href="#cb31-41" aria-hidden="true" tabindex="-1"></a>            </span>
<span id="cb31-42"><a href="#cb31-42" aria-hidden="true" tabindex="-1"></a>        <span class="cf">return</span> <span class="bu">sorted</span>(<span class="va">self</span>._df.<span class="bu">filter</span>(</span>
<span id="cb31-43"><a href="#cb31-43" aria-hidden="true" tabindex="-1"></a>            pl.col(column).<span class="bu">str</span>.contains(search_pat)</span>
<span id="cb31-44"><a href="#cb31-44" aria-hidden="true" tabindex="-1"></a>        ).select(column).to_series().to_list())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
</section>
</section>
</section>
<section id="mapping-checks" class="level1">
<h1>Mapping Checks</h1>
<section id="snomed-coded" class="level2">
<h2 class="anchored" data-anchor-id="snomed-coded">✅ SNOMED coded?</h2>
<div id="cell-51" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:23:00.341265Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:23:00.340972Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:23:00.453438Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:23:00.452450Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:23:00.341237Z&quot;}" data-trusted="true" data-execution_count="19">
<div class="sourceCode cell-code" id="cb32"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb32-1"><a href="#cb32-1" aria-hidden="true" tabindex="-1"></a>df.regex.search(pattern<span class="op">=</span><span class="st">"SNOMEDCT"</span>, column<span class="op">=</span><span class="st">'text_relationships'</span>)[<span class="dv">0</span>:<span class="dv">10</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="19">
<pre><code>['2-3 finger cutaneous syndactyly can be coded with SNOMEDCT_US 205139009, UMLS C0432055',
 '2-3 toe cutaneous syndactyly can be coded with SNOMEDCT_US 205145001, UMLS C0432040',
 '2-3 toe syndactyly can be coded with SNOMEDCT_US 205145001, UMLS C0432040',
 '3-m syndrome can be coded with SNOMEDCT_US 702342007, UMLS C1848862, UMLS C3280146',
 '3-methylglutaconic aciduria can be coded with SNOMEDCT_US 237950009, UMLS C3696376',
 '4-layered lissencephaly can be coded with SNOMEDCT_US 253147000, UMLS C0431375',
 'aagenaes syndrome can be coded with SNOMEDCT_US 28724005, UMLS C0268314',
 'abdominal aortic aneurysm can be coded with SNOMEDCT_US 155422008, UMLS C0162871',
 'abdominal colic can be coded with SNOMEDCT_US 9991008, UMLS C0232488',
 'abdominal distention can be coded with SNOMEDCT_US 41931001, SNOMEDCT_US 60728008, UMLS C0000731']</code></pre>
</div>
</div>
<div id="cell-52" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:23:00.455058Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:23:00.454696Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:23:00.469953Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:23:00.469076Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:23:00.455027Z&quot;}" data-trusted="true" data-execution_count="20">
<div class="sourceCode cell-code" id="cb34"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb34-1"><a href="#cb34-1" aria-hidden="true" tabindex="-1"></a><span class="bu">len</span>(df.regex.search(pattern<span class="op">=</span><span class="st">"SNOMEDCT"</span>, column<span class="op">=</span><span class="st">'text_relationships'</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="20">
<pre><code>8033</code></pre>
</div>
</div>
<div id="cell-53" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:23:00.471383Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:23:00.471067Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:23:00.476850Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:23:00.475826Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:23:00.471353Z&quot;}" data-trusted="true" data-execution_count="21">
<div class="sourceCode cell-code" id="cb36"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb36-1"><a href="#cb36-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Not a pure function but good for quick check</span></span>
<span id="cb36-2"><a href="#cb36-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> get_ontology_counts(search_term):</span>
<span id="cb36-3"><a href="#cb36-3" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'</span><span class="sc">{</span>search_term<span class="sc">}</span><span class="ss"> in Combined: </span><span class="sc">{</span>(<span class="bu">len</span>(df.regex.search(search_term, column<span class="op">=</span><span class="st">"text_relationships"</span>)))<span class="sc">}</span><span class="ss">'</span>)</span>
<span id="cb36-4"><a href="#cb36-4" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="st">"-"</span> <span class="op">*</span> <span class="dv">30</span>)</span>
<span id="cb36-5"><a href="#cb36-5" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'</span><span class="sc">{</span>search_term<span class="sc">}</span><span class="ss"> in HPO: </span><span class="sc">{</span>(<span class="bu">len</span>(hpo_df.regex.search(search_term, column<span class="op">=</span><span class="st">"text_relationships"</span>)))<span class="sc">}</span><span class="ss">'</span>)</span>
<span id="cb36-6"><a href="#cb36-6" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'</span><span class="sc">{</span>search_term<span class="sc">}</span><span class="ss"> in MONDO: </span><span class="sc">{</span>(<span class="bu">len</span>(mondo_df.regex.search(search_term, column<span class="op">=</span><span class="st">"text_relationships"</span>)))<span class="sc">}</span><span class="ss">'</span>)</span>
<span id="cb36-7"><a href="#cb36-7" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(<span class="ss">f'</span><span class="sc">{</span>search_term<span class="sc">}</span><span class="ss"> in DO: </span><span class="sc">{</span>(<span class="bu">len</span>(do_df.regex.search(search_term, column<span class="op">=</span><span class="st">"text_relationships"</span>)))<span class="sc">}</span><span class="ss">'</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-54" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:23:00.478278Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:23:00.477991Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:23:00.511491Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:23:00.510470Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:23:00.478248Z&quot;}" data-trusted="true" data-execution_count="22">
<div class="sourceCode cell-code" id="cb37"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb37-1"><a href="#cb37-1" aria-hidden="true" tabindex="-1"></a>get_ontology_counts(<span class="st">"SNOMED"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>SNOMED in Combined: 8033
------------------------------
SNOMED in HPO: 3448
SNOMED in MONDO: 0
SNOMED in DO: 4645</code></pre>
</div>
</div>
</section>
<section id="icd10-coded" class="level2">
<h2 class="anchored" data-anchor-id="icd10-coded">✅ ICD10 coded?</h2>
<div id="cell-56" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:23:00.513386Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:23:00.512848Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:23:00.532241Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:23:00.531266Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:23:00.513340Z&quot;}" data-trusted="true" data-execution_count="23">
<div class="sourceCode cell-code" id="cb39"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb39-1"><a href="#cb39-1" aria-hidden="true" tabindex="-1"></a>get_ontology_counts(<span class="st">"ICD10"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>ICD10 in Combined: 4996
------------------------------
ICD10 in HPO: 0
ICD10 in MONDO: 1637
ICD10 in DO: 3365</code></pre>
</div>
</div>
</section>
<section id="umls-coded" class="level2">
<h2 class="anchored" data-anchor-id="umls-coded">✅ UMLS coded?</h2>
<div id="cell-58" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:23:00.535029Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:23:00.534668Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:23:00.587982Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:23:00.586914Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:23:00.534998Z&quot;}" data-trusted="true" data-execution_count="24">
<div class="sourceCode cell-code" id="cb41"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb41-1"><a href="#cb41-1" aria-hidden="true" tabindex="-1"></a>get_ontology_counts(<span class="st">"UMLS"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>UMLS in Combined: 37201
------------------------------
UMLS in HPO: 11467
UMLS in MONDO: 20642
UMLS in DO: 6166</code></pre>
</div>
</div>
<div id="cell-59" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:23:00.589731Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:23:00.589289Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:23:00.619881Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:23:00.618951Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:23:00.589665Z&quot;}" data-trusted="true" data-execution_count="25">
<div class="sourceCode cell-code" id="cb43"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb43-1"><a href="#cb43-1" aria-hidden="true" tabindex="-1"></a>df.regex.search(<span class="st">"UMLS"</span>, column<span class="op">=</span><span class="st">'text_relationships'</span>)[<span class="dv">0</span>:<span class="dv">5</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="25">
<pre><code>['1-2 finger cutaneous syndactyly can be coded with UMLS C4023732',
 '1-2 toe complete cutaneous syndactyly can be coded with UMLS C4025140',
 '1-2 toe syndactyly can be coded with UMLS C4023726',
 '1-3 finger cutaneous syndactyly can be coded with UMLS C4023730',
 '1-3 toe syndactyly can be coded with UMLS C4025774']</code></pre>
</div>
</div>
</section>
<section id="check-stroke" class="level2">
<h2 class="anchored" data-anchor-id="check-stroke">✅ Check: “stroke”</h2>
<div id="cell-61" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:23:00.621577Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:23:00.621177Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:23:00.632410Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:23:00.631484Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:23:00.621531Z&quot;}" data-trusted="true" data-execution_count="26">
<div class="sourceCode cell-code" id="cb45"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb45-1"><a href="#cb45-1" aria-hidden="true" tabindex="-1"></a>df.regex.search(<span class="st">"stroke"</span>, column<span class="op">=</span><span class="st">"text_relationships"</span>)[<span class="dv">0</span>:<span class="dv">10</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="26">
<pre><code>['anterior spinal artery stroke can be coded with UMLS C2931608',
 'anterior spinal artery stroke is a arterial disorder',
 'anterior spinal artery stroke is a spinal cord ischemia',
 'autosomal recessive leukoencephalopathy-ischemic stroke-retinitis pigmentosa syndrome can be coded with UMLS C4749919',
 'autosomal recessive leukoencephalopathy-ischemic stroke-retinitis pigmentosa syndrome is a hereditary disease',
 'autosomal recessive leukoencephalopathy-ischemic stroke-retinitis pigmentosa syndrome is a syndromic disease',
 'cathepsin a-related arteriopathy-strokes-leukoencephalopathy can be coded with ICD10CM I67.8, UMLS C5680354',
 'cathepsin a-related arteriopathy-strokes-leukoencephalopathy is a cerebrovascular disorder',
 'cathepsin a-related arteriopathy-strokes-leukoencephalopathy is a hereditary neurological disease',
 'ischemic stroke can be coded with SNOMEDCT_US 422504002, UMLS C0948008']</code></pre>
</div>
</div>
</section>
<section id="check-moyamoya-disease" class="level2">
<h2 class="anchored" data-anchor-id="check-moyamoya-disease">✅ Check: “Moyamoya disease”</h2>
<div id="cell-63" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:23:00.634478Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:23:00.633732Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:23:00.645079Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:23:00.644095Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:23:00.634425Z&quot;}" data-trusted="true" data-execution_count="27">
<div class="sourceCode cell-code" id="cb47"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb47-1"><a href="#cb47-1" aria-hidden="true" tabindex="-1"></a>df.regex.search(<span class="st">"moyamoya"</span>, column<span class="op">=</span><span class="st">'text_relationships'</span>)[<span class="dv">10</span>:<span class="dv">20</span>]</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="27">
<pre><code>['moyamoya disease 5 can be coded with UMLS C3279690',
 'moyamoya disease 5 is a moyamoya disease',
 'moyamoya disease 7 can be coded with UMLS C5882748',
 'moyamoya disease 7 is a moyamoya disease',
 'moyamoya disease can be coded with ICD10CM I67.5, SNOMEDCT_US 69116000, UMLS C0026654',
 'moyamoya disease can be coded with UMLS C0026654',
 'moyamoya disease is a cerebral arterial disease',
 'moyamoya disease is a hereditary neurological disease',
 'moyamoya disease with early-onset achalasia can be coded with UMLS C3810403',
 'moyamoya disease with early-onset achalasia is a digestive system disorder']</code></pre>
</div>
</div>
</section>
</section>
<section id="token-number-rough-check" class="level1">
<h1>Token Number Rough check</h1>
<p>Note these are underestimates, since it just counts spaces.</p>
<section id="all-mondo-do-hpo" class="level2">
<h2 class="anchored" data-anchor-id="all-mondo-do-hpo">📈 All MONDO + DO + HPO</h2>
<div id="cell-66" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T09:20:54.764653Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T09:20:54.764365Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T09:20:54.948358Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T09:20:54.947363Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T09:20:54.764625Z&quot;}" data-trusted="true" data-execution_count="28">
<div class="sourceCode cell-code" id="cb49"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb49-1"><a href="#cb49-1" aria-hidden="true" tabindex="-1"></a><span class="bu">len</span>((<span class="st">". "</span>.join(<span class="bu">list</span>(<span class="bu">set</span>(relationships_from_mondo <span class="op">+</span> relationships_from_do <span class="op">+</span> relationships_from_hpo)))).split())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="28">
<pre><code>998762</code></pre>
</div>
</div>
</section>
<section id="mondo-1" class="level2">
<h2 class="anchored" data-anchor-id="mondo-1">📈 MONDO</h2>
<div id="cell-68" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T09:20:54.950192Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T09:20:54.949852Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T09:20:55.034006Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T09:20:55.032828Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T09:20:54.950162Z&quot;}" data-trusted="true" data-execution_count="29">
<div class="sourceCode cell-code" id="cb51"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb51-1"><a href="#cb51-1" aria-hidden="true" tabindex="-1"></a><span class="bu">len</span>((<span class="st">". "</span>.join(<span class="bu">list</span>(<span class="bu">set</span>(relationships_from_mondo)))).split())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="29">
<pre><code>502867</code></pre>
</div>
</div>
</section>
<section id="do" class="level2">
<h2 class="anchored" data-anchor-id="do">📈 DO</h2>
<div id="cell-70" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T09:20:55.035846Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T09:20:55.035411Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T09:20:55.072481Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T09:20:55.071364Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T09:20:55.035801Z&quot;}" data-trusted="true" data-execution_count="30">
<div class="sourceCode cell-code" id="cb53"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb53-1"><a href="#cb53-1" aria-hidden="true" tabindex="-1"></a><span class="bu">len</span>((<span class="st">". "</span>.join(<span class="bu">list</span>(<span class="bu">set</span>(relationships_from_do)))).split())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="30">
<pre><code>178064</code></pre>
</div>
</div>
</section>
<section id="hpo-1" class="level2">
<h2 class="anchored" data-anchor-id="hpo-1">📈 HPO</h2>
<div id="cell-72" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T09:20:55.074727Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T09:20:55.074292Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T09:20:55.136183Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T09:20:55.135134Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T09:20:55.074679Z&quot;}" data-trusted="true" data-execution_count="31">
<div class="sourceCode cell-code" id="cb55"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb55-1"><a href="#cb55-1" aria-hidden="true" tabindex="-1"></a><span class="bu">len</span>((<span class="st">". "</span>.join(<span class="bu">list</span>(<span class="bu">set</span>(relationships_from_hpo)))).split())</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="31">
<pre><code>357317</code></pre>
</div>
</div>
<div id="cell-73" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:24:36.525835Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:24:36.525376Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:24:36.530654Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:24:36.529666Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:24:36.525796Z&quot;}" data-trusted="true" data-execution_count="32">
<div class="sourceCode cell-code" id="cb57"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb57-1"><a href="#cb57-1" aria-hidden="true" tabindex="-1"></a>df <span class="op">=</span> <span class="va">None</span></span>
<span id="cb57-2"><a href="#cb57-2" aria-hidden="true" tabindex="-1"></a>hpo_df <span class="op">=</span> <span class="va">None</span></span>
<span id="cb57-3"><a href="#cb57-3" aria-hidden="true" tabindex="-1"></a>mondo_df <span class="op">=</span> <span class="va">None</span></span>
<span id="cb57-4"><a href="#cb57-4" aria-hidden="true" tabindex="-1"></a>do_df <span class="op">=</span> <span class="va">None</span></span>
<span id="cb57-5"><a href="#cb57-5" aria-hidden="true" tabindex="-1"></a>onto <span class="op">=</span> <span class="va">None</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<hr>
</section>
</section>
<section id="finally-google-gemini-work" class="level1">
<h1>Finally, Google Gemini Work</h1>
</section>
<section id="kaggle-seems-to-have-trouble-running-prompts-500k-tokens-it-appears" class="level1">
<h1>⚠️ KAGGLE Seems to have trouble running PROMPTS &gt;500k tokens (?) IT APPEARS</h1>
<p>I am resorting to single-ontology mapping.</p>
<p><strong>Generally, the multiple ontology approach prevents code hallucination, but I’ll switch between HPO and Disease Ontology because of the limitations of Kaggle Notebooks.</strong></p>
<div id="cell-76" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:25:14.954154Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:25:14.953777Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:25:15.242421Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:25:15.241556Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:25:14.954122Z&quot;}" data-trusted="true" data-execution_count="33">
<div class="sourceCode cell-code" id="cb58"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb58-1"><a href="#cb58-1" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> os</span>
<span id="cb58-2"><a href="#cb58-2" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> google.generativeai <span class="im">as</span> genai</span>
<span id="cb58-3"><a href="#cb58-3" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> google.generativeai <span class="im">import</span> caching</span>
<span id="cb58-4"><a href="#cb58-4" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> google.generativeai.types <span class="im">import</span> RequestOptions</span>
<span id="cb58-5"><a href="#cb58-5" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> google.api_core <span class="im">import</span> retry</span>
<span id="cb58-6"><a href="#cb58-6" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> time <span class="im">import</span> sleep</span>
<span id="cb58-7"><a href="#cb58-7" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> dotenv <span class="im">import</span> load_dotenv</span>
<span id="cb58-8"><a href="#cb58-8" aria-hidden="true" tabindex="-1"></a><span class="im">import</span> json</span>
<span id="cb58-9"><a href="#cb58-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-10"><a href="#cb58-10" aria-hidden="true" tabindex="-1"></a>load_dotenv()</span>
<span id="cb58-11"><a href="#cb58-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-12"><a href="#cb58-12" aria-hidden="true" tabindex="-1"></a>API_KEY <span class="op">=</span> os.environ.get(<span class="st">'GEMINI_API_KEY'</span>)</span>
<span id="cb58-13"><a href="#cb58-13" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb58-14"><a href="#cb58-14" aria-hidden="true" tabindex="-1"></a>genai.configure(api_key<span class="op">=</span>API_KEY)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-77" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:25:16.437213Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:25:16.436394Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:25:17.278328Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:25:17.277200Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:25:16.437172Z&quot;}" data-trusted="true" data-execution_count="74">
<div class="sourceCode cell-code" id="cb59"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb59-1"><a href="#cb59-1" aria-hidden="true" tabindex="-1"></a><span class="co"># If you need to clear the cache completely:</span></span>
<span id="cb59-2"><a href="#cb59-2" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> clear_cache():</span>
<span id="cb59-3"><a href="#cb59-3" aria-hidden="true" tabindex="-1"></a>    <span class="cf">for</span> c <span class="kw">in</span> caching.CachedContent.<span class="bu">list</span>():</span>
<span id="cb59-4"><a href="#cb59-4" aria-hidden="true" tabindex="-1"></a>        c.delete()</span>
<span id="cb59-5"><a href="#cb59-5" aria-hidden="true" tabindex="-1"></a>clear_cache()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-78" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:25:19.138171Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:25:19.137765Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:25:19.428034Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:25:19.427155Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:25:19.138135Z&quot;}" data-trusted="true" data-execution_count="75">
<div class="sourceCode cell-code" id="cb60"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb60-1"><a href="#cb60-1" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span> c <span class="kw">in</span> caching.CachedContent.<span class="bu">list</span>():</span>
<span id="cb60-2"><a href="#cb60-2" aria-hidden="true" tabindex="-1"></a>    <span class="bu">print</span>(c)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="application-structured-consultation-reports" class="level1">
<h1>APPLICATION: Structured Consultation Reports</h1>
<section id="the-prompt" class="level2">
<h2 class="anchored" data-anchor-id="the-prompt">The Prompt</h2>
<p>Parse any content clinical information, digest it and encode it with standard terminology codes.</p>
<div id="cell-80" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:25:24.551072Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:25:24.550629Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:25:38.185781Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:25:38.184837Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:25:24.551035Z&quot;}" data-trusted="true" data-execution_count="58">
<div class="sourceCode cell-code" id="cb61"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb61-1"><a href="#cb61-1" aria-hidden="true" tabindex="-1"></a>BASE_INSTRUCTIONS <span class="op">=</span> <span class="st">"""</span></span>
<span id="cb61-2"><a href="#cb61-2" aria-hidden="true" tabindex="-1"></a><span class="st">Generate structured consultation reports from clinical inputs. </span></span>
<span id="cb61-3"><a href="#cb61-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-4"><a href="#cb61-4" aria-hidden="true" tabindex="-1"></a><span class="st">Use these status indicators when appropriate:</span></span>
<span id="cb61-5"><a href="#cb61-5" aria-hidden="true" tabindex="-1"></a><span class="st">NEGATION (use both): </span></span>
<span id="cb61-6"><a href="#cb61-6" aria-hidden="true" tabindex="-1"></a><span class="st">  ❌ finding + "no/absent/without [term]"</span></span>
<span id="cb61-7"><a href="#cb61-7" aria-hidden="true" tabindex="-1"></a><span class="st">SEVERITY: </span></span>
<span id="cb61-8"><a href="#cb61-8" aria-hidden="true" tabindex="-1"></a><span class="st">  🔴 severe/critical</span></span>
<span id="cb61-9"><a href="#cb61-9" aria-hidden="true" tabindex="-1"></a><span class="st">  🟡 moderate/concerning</span></span>
<span id="cb61-10"><a href="#cb61-10" aria-hidden="true" tabindex="-1"></a><span class="st">  🟢 mild/stable</span></span>
<span id="cb61-11"><a href="#cb61-11" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-12"><a href="#cb61-12" aria-hidden="true" tabindex="-1"></a><span class="st">INPUT CLASSIFICATION:</span></span>
<span id="cb61-13"><a href="#cb61-13" aria-hidden="true" tabindex="-1"></a><span class="st">- Clearly state source type: Patient text or question, clinical text, radiology image, pathology image, or photo of a lesion, etc.</span></span>
<span id="cb61-14"><a href="#cb61-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-15"><a href="#cb61-15" aria-hidden="true" tabindex="-1"></a><span class="st">REPORT FORMAT:</span></span>
<span id="cb61-16"><a href="#cb61-16" aria-hidden="true" tabindex="-1"></a><span class="st">### Source</span></span>
<span id="cb61-17"><a href="#cb61-17" aria-hidden="true" tabindex="-1"></a><span class="st">[One line describing input type and key context]</span></span>
<span id="cb61-18"><a href="#cb61-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-19"><a href="#cb61-19" aria-hidden="true" tabindex="-1"></a><span class="st">### Impression</span></span>
<span id="cb61-20"><a href="#cb61-20" aria-hidden="true" tabindex="-1"></a><span class="st">[2-3 lines maximum capturing key findings/diagnosis]</span></span>
<span id="cb61-21"><a href="#cb61-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-22"><a href="#cb61-22" aria-hidden="true" tabindex="-1"></a><span class="st">### Supporting Evidence</span></span>
<span id="cb61-23"><a href="#cb61-23" aria-hidden="true" tabindex="-1"></a><span class="st">- Primary findings, each with:</span></span>
<span id="cb61-24"><a href="#cb61-24" aria-hidden="true" tabindex="-1"></a><span class="st">  - Cached ontology and knowledge base mapping</span></span>
<span id="cb61-25"><a href="#cb61-25" aria-hidden="true" tabindex="-1"></a><span class="st">  - Only from cached ontology, find codes from terminologies SNOMEDCT_US, ICD10CM, and/or UMLS when present in the cached ontology: Format as `code`.</span></span>
<span id="cb61-26"><a href="#cb61-26" aria-hidden="true" tabindex="-1"></a><span class="st">  - Supporting details from input</span></span>
<span id="cb61-27"><a href="#cb61-27" aria-hidden="true" tabindex="-1"></a><span class="st">- Add Markdown table for identified concepts </span></span>
<span id="cb61-28"><a href="#cb61-28" aria-hidden="true" tabindex="-1"></a><span class="st">    STRING | TERMINOLOGY | CODE</span></span>
<span id="cb61-29"><a href="#cb61-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-30"><a href="#cb61-30" aria-hidden="true" tabindex="-1"></a><span class="st">### Relevant Relationships</span></span>
<span id="cb61-31"><a href="#cb61-31" aria-hidden="true" tabindex="-1"></a><span class="st">- Key ontological associations:</span></span>
<span id="cb61-32"><a href="#cb61-32" aria-hidden="true" tabindex="-1"></a><span class="st">  - Direct phenotype/disease links</span></span>
<span id="cb61-33"><a href="#cb61-33" aria-hidden="true" tabindex="-1"></a><span class="st">  - Inheritance patterns if applicable</span></span>
<span id="cb61-34"><a href="#cb61-34" aria-hidden="true" tabindex="-1"></a><span class="st">  - System-level relationships</span></span>
<span id="cb61-35"><a href="#cb61-35" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-36"><a href="#cb61-36" aria-hidden="true" tabindex="-1"></a><span class="st">FORMAT RULES:</span></span>
<span id="cb61-37"><a href="#cb61-37" aria-hidden="true" tabindex="-1"></a><span class="st">- Use `backticks` for ontology terms</span></span>
<span id="cb61-38"><a href="#cb61-38" aria-hidden="true" tabindex="-1"></a><span class="st">- **Bold** for key matches/findings</span></span>
<span id="cb61-39"><a href="#cb61-39" aria-hidden="true" tabindex="-1"></a><span class="st">- Group related findings</span></span>
<span id="cb61-40"><a href="#cb61-40" aria-hidden="true" tabindex="-1"></a><span class="st">- Brief, evidence-based statements</span></span>
<span id="cb61-41"><a href="#cb61-41" aria-hidden="true" tabindex="-1"></a><span class="st">- Hyperlink SNOMEDCT_US code as: [REPLACE WITH CODE](https://browser.ihtsdotools.org/?perspective=full&amp;conceptId1=[REPLACE WITH CODE]&amp;edition=MAIN/SNOMEDCT-US/2024-09-01&amp;release=&amp;languages=en)</span></span>
<span id="cb61-42"><a href="#cb61-42" aria-hidden="true" tabindex="-1"></a><span class="st">- Hyperlink UMLS code as: [REPLACE WITH CODE](https://uts.nlm.nih.gov/uts/umls/searchResults?searchString=[REPLACE WITH CODE]) </span></span>
<span id="cb61-43"><a href="#cb61-43" aria-hidden="true" tabindex="-1"></a><span class="st"># """</span></span>
<span id="cb61-44"><a href="#cb61-44" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-45"><a href="#cb61-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-46"><a href="#cb61-46" aria-hidden="true" tabindex="-1"></a><span class="co"># cache_hpo = caching.CachedContent.create(</span></span>
<span id="cb61-47"><a href="#cb61-47" aria-hidden="true" tabindex="-1"></a><span class="co">#     model="models/gemini-1.5-pro-002",</span></span>
<span id="cb61-48"><a href="#cb61-48" aria-hidden="true" tabindex="-1"></a><span class="co">#     display_name='Human Phenotype Ontology',</span></span>
<span id="cb61-49"><a href="#cb61-49" aria-hidden="true" tabindex="-1"></a><span class="co">#     ttl=60*60,</span></span>
<span id="cb61-50"><a href="#cb61-50" aria-hidden="true" tabindex="-1"></a><span class="co">#     system_instruction=f"""You are a clinical consultation system that analyzes inputs and provides structured reports</span></span>
<span id="cb61-51"><a href="#cb61-51" aria-hidden="true" tabindex="-1"></a><span class="co">#     using ontology-backed reasoning. Use only cached relationships for interpretations.</span></span>
<span id="cb61-52"><a href="#cb61-52" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb61-53"><a href="#cb61-53" aria-hidden="true" tabindex="-1"></a><span class="co">#     Core Functions:</span></span>
<span id="cb61-54"><a href="#cb61-54" aria-hidden="true" tabindex="-1"></a><span class="co">#     - Input classification</span></span>
<span id="cb61-55"><a href="#cb61-55" aria-hidden="true" tabindex="-1"></a><span class="co">#     - Key finding identification</span></span>
<span id="cb61-56"><a href="#cb61-56" aria-hidden="true" tabindex="-1"></a><span class="co">#     - Ontology-based reasoning</span></span>
<span id="cb61-57"><a href="#cb61-57" aria-hidden="true" tabindex="-1"></a><span class="co">#     - Evidence structuring</span></span>
<span id="cb61-58"><a href="#cb61-58" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb61-59"><a href="#cb61-59" aria-hidden="true" tabindex="-1"></a><span class="co">#     {BASE_INSTRUCTIONS}</span></span>
<span id="cb61-60"><a href="#cb61-60" aria-hidden="true" tabindex="-1"></a><span class="co">#     """,</span></span>
<span id="cb61-61"><a href="#cb61-61" aria-hidden="true" tabindex="-1"></a><span class="co">#     contents=". ".join(</span></span>
<span id="cb61-62"><a href="#cb61-62" aria-hidden="true" tabindex="-1"></a><span class="co">#         list(</span></span>
<span id="cb61-63"><a href="#cb61-63" aria-hidden="true" tabindex="-1"></a><span class="co">#             set(</span></span>
<span id="cb61-64"><a href="#cb61-64" aria-hidden="true" tabindex="-1"></a><span class="co">#                 relationships_from_hpo</span></span>
<span id="cb61-65"><a href="#cb61-65" aria-hidden="true" tabindex="-1"></a><span class="co">#             )</span></span>
<span id="cb61-66"><a href="#cb61-66" aria-hidden="true" tabindex="-1"></a><span class="co">#         )</span></span>
<span id="cb61-67"><a href="#cb61-67" aria-hidden="true" tabindex="-1"></a><span class="co">#     )</span></span>
<span id="cb61-68"><a href="#cb61-68" aria-hidden="true" tabindex="-1"></a><span class="co"># )</span></span>
<span id="cb61-69"><a href="#cb61-69" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-70"><a href="#cb61-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-71"><a href="#cb61-71" aria-hidden="true" tabindex="-1"></a>cache_all <span class="op">=</span> caching.CachedContent.create(</span>
<span id="cb61-72"><a href="#cb61-72" aria-hidden="true" tabindex="-1"></a>    model<span class="op">=</span><span class="st">"models/gemini-1.5-pro-002"</span>,</span>
<span id="cb61-73"><a href="#cb61-73" aria-hidden="true" tabindex="-1"></a>    display_name<span class="op">=</span><span class="st">'HPO + DO'</span>,</span>
<span id="cb61-74"><a href="#cb61-74" aria-hidden="true" tabindex="-1"></a>    ttl<span class="op">=</span><span class="dv">60</span><span class="op">*</span><span class="dv">60</span>,</span>
<span id="cb61-75"><a href="#cb61-75" aria-hidden="true" tabindex="-1"></a>    system_instruction<span class="op">=</span><span class="ss">f"""You are a clinical consultation system that analyzes inputs and provides structured reports</span></span>
<span id="cb61-76"><a href="#cb61-76" aria-hidden="true" tabindex="-1"></a><span class="ss">    using ontology-backed reasoning. Use only cached relationships for interpretations.</span></span>
<span id="cb61-77"><a href="#cb61-77" aria-hidden="true" tabindex="-1"></a><span class="ss">    </span></span>
<span id="cb61-78"><a href="#cb61-78" aria-hidden="true" tabindex="-1"></a><span class="ss">    Core Functions:</span></span>
<span id="cb61-79"><a href="#cb61-79" aria-hidden="true" tabindex="-1"></a><span class="ss">    - Input classification</span></span>
<span id="cb61-80"><a href="#cb61-80" aria-hidden="true" tabindex="-1"></a><span class="ss">    - Key finding identification</span></span>
<span id="cb61-81"><a href="#cb61-81" aria-hidden="true" tabindex="-1"></a><span class="ss">    - Ontology-based reasoning</span></span>
<span id="cb61-82"><a href="#cb61-82" aria-hidden="true" tabindex="-1"></a><span class="ss">    - Evidence structuring</span></span>
<span id="cb61-83"><a href="#cb61-83" aria-hidden="true" tabindex="-1"></a><span class="ss">  </span></span>
<span id="cb61-84"><a href="#cb61-84" aria-hidden="true" tabindex="-1"></a><span class="ss">    </span><span class="sc">{</span>BASE_INSTRUCTIONS<span class="sc">}</span></span>
<span id="cb61-85"><a href="#cb61-85" aria-hidden="true" tabindex="-1"></a><span class="ss">    """</span>,</span>
<span id="cb61-86"><a href="#cb61-86" aria-hidden="true" tabindex="-1"></a>    contents<span class="op">=</span><span class="st">". "</span>.join(</span>
<span id="cb61-87"><a href="#cb61-87" aria-hidden="true" tabindex="-1"></a>        <span class="bu">list</span>(</span>
<span id="cb61-88"><a href="#cb61-88" aria-hidden="true" tabindex="-1"></a>            <span class="bu">set</span>(</span>
<span id="cb61-89"><a href="#cb61-89" aria-hidden="true" tabindex="-1"></a>                relationships_from_hpo</span>
<span id="cb61-90"><a href="#cb61-90" aria-hidden="true" tabindex="-1"></a>                <span class="op">+</span> relationships_from_do</span>
<span id="cb61-91"><a href="#cb61-91" aria-hidden="true" tabindex="-1"></a>            )</span>
<span id="cb61-92"><a href="#cb61-92" aria-hidden="true" tabindex="-1"></a>        )</span>
<span id="cb61-93"><a href="#cb61-93" aria-hidden="true" tabindex="-1"></a>    )</span>
<span id="cb61-94"><a href="#cb61-94" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb61-95"><a href="#cb61-95" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-96"><a href="#cb61-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb61-97"><a href="#cb61-97" aria-hidden="true" tabindex="-1"></a><span class="co"># cache_do = caching.CachedContent.create(</span></span>
<span id="cb61-98"><a href="#cb61-98" aria-hidden="true" tabindex="-1"></a><span class="co">#     model="models/gemini-1.5-pro-002",</span></span>
<span id="cb61-99"><a href="#cb61-99" aria-hidden="true" tabindex="-1"></a><span class="co">#     display_name='Disese Ontology',</span></span>
<span id="cb61-100"><a href="#cb61-100" aria-hidden="true" tabindex="-1"></a><span class="co">#     ttl=60*60*2,</span></span>
<span id="cb61-101"><a href="#cb61-101" aria-hidden="true" tabindex="-1"></a><span class="co">#     system_instruction=f"""You are a clinical consultation system that analyzes inputs and provides structured reports</span></span>
<span id="cb61-102"><a href="#cb61-102" aria-hidden="true" tabindex="-1"></a><span class="co">#     using ontology-backed reasoning. Use only cached relationships for interpretations.</span></span>
<span id="cb61-103"><a href="#cb61-103" aria-hidden="true" tabindex="-1"></a>    </span>
<span id="cb61-104"><a href="#cb61-104" aria-hidden="true" tabindex="-1"></a><span class="co">#     Core Functions:</span></span>
<span id="cb61-105"><a href="#cb61-105" aria-hidden="true" tabindex="-1"></a><span class="co">#     - Input classification</span></span>
<span id="cb61-106"><a href="#cb61-106" aria-hidden="true" tabindex="-1"></a><span class="co">#     - Key finding identification</span></span>
<span id="cb61-107"><a href="#cb61-107" aria-hidden="true" tabindex="-1"></a><span class="co">#     - Ontology-based reasoning</span></span>
<span id="cb61-108"><a href="#cb61-108" aria-hidden="true" tabindex="-1"></a><span class="co">#     - Evidence structuring</span></span>
<span id="cb61-109"><a href="#cb61-109" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb61-110"><a href="#cb61-110" aria-hidden="true" tabindex="-1"></a><span class="co">#     {BASE_INSTRUCTIONS}</span></span>
<span id="cb61-111"><a href="#cb61-111" aria-hidden="true" tabindex="-1"></a><span class="co">#     """,</span></span>
<span id="cb61-112"><a href="#cb61-112" aria-hidden="true" tabindex="-1"></a><span class="co">#     contents=". ".join(</span></span>
<span id="cb61-113"><a href="#cb61-113" aria-hidden="true" tabindex="-1"></a><span class="co">#         list(</span></span>
<span id="cb61-114"><a href="#cb61-114" aria-hidden="true" tabindex="-1"></a><span class="co">#             set(</span></span>
<span id="cb61-115"><a href="#cb61-115" aria-hidden="true" tabindex="-1"></a><span class="co">#                 relationships_from_do</span></span>
<span id="cb61-116"><a href="#cb61-116" aria-hidden="true" tabindex="-1"></a><span class="co">#             )</span></span>
<span id="cb61-117"><a href="#cb61-117" aria-hidden="true" tabindex="-1"></a><span class="co">#         )</span></span>
<span id="cb61-118"><a href="#cb61-118" aria-hidden="true" tabindex="-1"></a><span class="co">#     )</span></span>
<span id="cb61-119"><a href="#cb61-119" aria-hidden="true" tabindex="-1"></a><span class="co"># )</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<section id="markdown-output-displayer" class="level3">
<h3 class="anchored" data-anchor-id="markdown-output-displayer">Markdown Output Displayer</h3>
<div id="cell-82" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:25:38.187590Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:25:38.187270Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:25:38.192543Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:25:38.191396Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:25:38.187560Z&quot;}" data-trusted="true" data-execution_count="59">
<div class="sourceCode cell-code" id="cb62"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb62-1"><a href="#cb62-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> IPython.display <span class="im">import</span> display, Markdown</span>
<span id="cb62-2"><a href="#cb62-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb62-3"><a href="#cb62-3" aria-hidden="true" tabindex="-1"></a><span class="kw">def</span> render_markdown(markdown_text):</span>
<span id="cb62-4"><a href="#cb62-4" aria-hidden="true" tabindex="-1"></a>    display(Markdown(markdown_text))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="patient-chief-complaint" class="level2">
<h2 class="anchored" data-anchor-id="patient-chief-complaint">🎯 Patient Chief Complaint</h2>
<div id="cell-84" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T11:21:56.100443Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T11:21:56.100048Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T11:22:26.352237Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T11:22:26.351180Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T11:21:56.100408Z&quot;}" data-trusted="true" data-execution_count="60">
<div class="sourceCode cell-code" id="cb63"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb63-1"><a href="#cb63-1" aria-hidden="true" tabindex="-1"></a>user_input <span class="op">=</span> <span class="st">"""My belly hurts 10/10 sharp pain."""</span></span>
<span id="cb63-2"><a href="#cb63-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb63-3"><a href="#cb63-3" aria-hidden="true" tabindex="-1"></a>model_cached <span class="op">=</span> genai.GenerativeModel.from_cached_content(cached_content<span class="op">=</span>cache_all)</span>
<span id="cb63-4"><a href="#cb63-4" aria-hidden="true" tabindex="-1"></a>ontology_response <span class="op">=</span> model_cached.generate_content(</span>
<span id="cb63-5"><a href="#cb63-5" aria-hidden="true" tabindex="-1"></a>   <span class="ss">f"""Using ONLY the cached ontology relationships, provide a structured analysis report for the following input: </span><span class="sc">{</span>user_input<span class="sc">}</span></span>
<span id="cb63-6"><a href="#cb63-6" aria-hidden="true" tabindex="-1"></a><span class="ss">    """</span>,</span>
<span id="cb63-7"><a href="#cb63-7" aria-hidden="true" tabindex="-1"></a>    request_options<span class="op">=</span>{<span class="st">"timeout"</span>: <span class="dv">600</span>}</span>
<span id="cb63-8"><a href="#cb63-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb63-9"><a href="#cb63-9" aria-hidden="true" tabindex="-1"></a>render_markdown(<span class="ss">f"</span><span class="sc">{</span>ontology_response<span class="sc">.</span>text<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display cell-output-markdown">
<section id="source" class="level3">
<h3 class="anchored" data-anchor-id="source">Source</h3>
<p>Patient text describing severe abdominal pain.</p>
</section>
<section id="impression" class="level3">
<h3 class="anchored" data-anchor-id="impression">Impression</h3>
<p>❌ Appendicitis: Patient denies fever, nausea, or vomiting. 🔴 Severe abdominal pain of unspecified cause. 🟢 Additional symptoms should be collected to determine the cause of the pain.</p>
</section>
<section id="supporting-evidence" class="level3">
<h3 class="anchored" data-anchor-id="supporting-evidence">Supporting Evidence</h3>
<ul>
<li><strong>Abdominal pain</strong> present
<ul>
<li><code>Abdominal pain</code></li>
<li>SNOMEDCT_US: <a href="https://browser.ihtsdotools.org/?perspective=full&amp;conceptId1=74400008&amp;edition=MAIN/SNOMEDCT-US/2024-09-01&amp;release=&amp;languages=en">74400008</a></li>
<li>ICD10CM: R10.9</li>
<li>UMLS: <a href="https://uts.nlm.nih.gov/uts/umls/searchResults?searchString=C0000737">C0000737</a></li>
<li>Patient states: “My belly hurts”.</li>
</ul></li>
<li><strong>Sharp pain</strong> quality
<ul>
<li><code>Sharp pain</code> quality</li>
<li>Patient states: “sharp pain”.</li>
</ul></li>
<li>Pain severity <strong>10/10</strong>
<ul>
<li><code>Severe</code> pain</li>
<li>Patient states: “10/10 pain”.</li>
</ul></li>
</ul>
<table class="table table-sm table-striped small">
<thead>
<tr class="header">
<th>STRING</th>
<th>TERMINOLOGY</th>
<th>CODE</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Abdominal pain</td>
<td>SNOMEDCT_US</td>
<td>74400008</td>
</tr>
<tr class="even">
<td>Abdominal pain</td>
<td>ICD10CM</td>
<td>R10.9</td>
</tr>
<tr class="odd">
<td>Abdominal pain</td>
<td>UMLS</td>
<td>C0000737</td>
</tr>
</tbody>
</table>
</section>
<section id="relevant-relationships" class="level3">
<h3 class="anchored" data-anchor-id="relevant-relationships">Relevant Relationships</h3>
<ul>
<li><code>Abdominal pain</code> is a symptom.</li>
<li><code>Severe</code> pain is classified by pain intensity.</li>
<li><code>Sharp</code> characterizes pain quality.</li>
</ul>
</section>
</div>
</div>
</section>
<section id="patient-chatbot-log-1" class="level2">
<h2 class="anchored" data-anchor-id="patient-chatbot-log-1">🎯 Patient Chatbot Log #1</h2>
<div id="cell-86" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T12:26:05.768832Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:26:05.768429Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:26:41.101215Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:26:41.099989Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:26:05.768800Z&quot;}" data-trusted="true" data-execution_count="62">
<div class="sourceCode cell-code" id="cb64"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb64-1"><a href="#cb64-1" aria-hidden="true" tabindex="-1"></a>user_input <span class="op">=</span> <span class="st">"""</span></span>
<span id="cb64-2"><a href="#cb64-2" aria-hidden="true" tabindex="-1"></a><span class="st">    Hi AI assistant, I've been experiencing heart palpitations and dizziness for about 3 months. </span></span>
<span id="cb64-3"><a href="#cb64-3" aria-hidden="true" tabindex="-1"></a><span class="st">    Can you help me understand what this might be?</span></span>
<span id="cb64-4"><a href="#cb64-4" aria-hidden="true" tabindex="-1"></a><span class="st">    """</span></span>
<span id="cb64-5"><a href="#cb64-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb64-6"><a href="#cb64-6" aria-hidden="true" tabindex="-1"></a>model_cached <span class="op">=</span> genai.GenerativeModel.from_cached_content(cached_content<span class="op">=</span>cache_all)</span>
<span id="cb64-7"><a href="#cb64-7" aria-hidden="true" tabindex="-1"></a>ontology_response <span class="op">=</span> model_cached.generate_content(</span>
<span id="cb64-8"><a href="#cb64-8" aria-hidden="true" tabindex="-1"></a>   <span class="ss">f"""Using ONLY the cached ontology relationships, provide a structured analysis report for the following input: </span><span class="sc">{</span>user_input<span class="sc">}</span></span>
<span id="cb64-9"><a href="#cb64-9" aria-hidden="true" tabindex="-1"></a><span class="ss">    """</span></span>
<span id="cb64-10"><a href="#cb64-10" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb64-11"><a href="#cb64-11" aria-hidden="true" tabindex="-1"></a>render_markdown(<span class="ss">f"</span><span class="sc">{</span>ontology_response<span class="sc">.</span>text<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display cell-output-markdown">
<section id="source-1" class="level3">
<h3 class="anchored" data-anchor-id="source-1">Source</h3>
<p>Patient question describing heart palpitations and dizziness.</p>
</section>
<section id="impression-1" class="level3">
<h3 class="anchored" data-anchor-id="impression-1">Impression</h3>
<p>🟡 <strong>Concerning for a cardiac arrhythmia.</strong> Further evaluation is needed to determine the specific type and cause of the arrhythmia. Dizziness may be related to the palpitations or indicate an additional condition.</p>
</section>
<section id="supporting-evidence-1" class="level3">
<h3 class="anchored" data-anchor-id="supporting-evidence-1">Supporting Evidence</h3>
<ul>
<li><strong>Heart palpitations:</strong>
<ul>
<li><code>Heart palpitation</code></li>
<li>SNOMEDCT_US: <a href="https://browser.ihtsdotools.org/?perspective=full&amp;conceptId1=80313002&amp;edition=MAIN/SNOMEDCT-US/2024-09-01&amp;release=&amp;languages=en">80313002</a> | ICD10CM R00.2</li>
<li>UMLS: <a href="https://uts.nlm.nih.gov/uts/umls/searchResults?searchString=C0030252">C0030252</a></li>
<li>Described as occurring for three months</li>
</ul></li>
<li><strong>Dizziness:</strong>
<ul>
<li><code>Dizziness</code></li>
<li>SNOMEDCT_US: <a href="https://browser.ihtsdotools.org/?perspective=full&amp;conceptId1=404640003&amp;edition=MAIN/SNOMEDCT-US/2024-09-01&amp;release=&amp;languages=en">404640003</a></li>
<li>UMLS: <a href="https://uts.nlm.nih.gov/uts/umls/searchResults?searchString=C0042571">C0042571</a>, <a href="https://uts.nlm.nih.gov/uts/umls/searchResults?searchString=C0012833">C0012833</a> | ICD10CM R42</li>
<li>Patient reports experiencing dizziness for three months</li>
</ul></li>
</ul>
<table class="table table-sm table-striped small">
<thead>
<tr class="header">
<th>STRING</th>
<th>TERMINOLOGY</th>
<th>CODE</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Heart palpitation</td>
<td>SNOMEDCT_US</td>
<td>80313002</td>
</tr>
<tr class="even">
<td>Heart palpitation</td>
<td>ICD10CM</td>
<td>R00.2</td>
</tr>
<tr class="odd">
<td>Heart palpitation</td>
<td>UMLS</td>
<td>C0030252</td>
</tr>
<tr class="even">
<td>Dizziness</td>
<td>SNOMEDCT_US</td>
<td>404640003</td>
</tr>
<tr class="odd">
<td>Dizziness</td>
<td>ICD10CM</td>
<td>R42</td>
</tr>
<tr class="even">
<td>Dizziness</td>
<td>UMLS</td>
<td>C0042571</td>
</tr>
<tr class="odd">
<td>Dizziness</td>
<td>UMLS</td>
<td>C0012833</td>
</tr>
</tbody>
</table>
</section>
<section id="relevant-relationships-1" class="level3">
<h3 class="anchored" data-anchor-id="relevant-relationships-1">Relevant Relationships</h3>
<ul>
<li><code>Heart palpitation</code> is found in <code>cardiac arrhythmia.</code></li>
<li><code>Dizziness</code> can be associated with some <code>cardiac arrhythmia</code>s.<br>
</li>
<li><code>Dizziness</code> can be a symptom of other conditions and may not be directly related to <code>heart palpitation</code>.</li>
</ul>
</section>
</div>
</div>
</section>
<section id="patient-chatbot-log-2" class="level2">
<h2 class="anchored" data-anchor-id="patient-chatbot-log-2">🎯 Patient Chatbot Log #2</h2>
<div id="cell-88" class="cell" data-execution="{&quot;iopub.status.busy&quot;:&quot;2024-12-01T12:25:38.228829Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T12:25:38.229433Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T12:25:38.229183Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T12:25:38.229154Z&quot;}" data-trusted="true" data-execution_count="63">
<div class="sourceCode cell-code" id="cb65"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb65-1"><a href="#cb65-1" aria-hidden="true" tabindex="-1"></a>user_input <span class="op">=</span> <span class="st">"""I have diabetes and high blood pressure. For the last 3 days I've been short of breath when moving around, need extra pillows to sleep, and my legs are swollen. I gained 4 pounds this week. What could this mean?"""</span></span>
<span id="cb65-2"><a href="#cb65-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb65-3"><a href="#cb65-3" aria-hidden="true" tabindex="-1"></a>model_cached <span class="op">=</span> genai.GenerativeModel.from_cached_content(cached_content<span class="op">=</span>cache_all)</span>
<span id="cb65-4"><a href="#cb65-4" aria-hidden="true" tabindex="-1"></a>ontology_response <span class="op">=</span> model_cached.generate_content(</span>
<span id="cb65-5"><a href="#cb65-5" aria-hidden="true" tabindex="-1"></a>   <span class="ss">f"""Using ONLY the cached ontology relationships, provide a structured analysis report for the following input: </span><span class="sc">{</span>user_input<span class="sc">}</span></span>
<span id="cb65-6"><a href="#cb65-6" aria-hidden="true" tabindex="-1"></a><span class="ss">    """</span></span>
<span id="cb65-7"><a href="#cb65-7" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb65-8"><a href="#cb65-8" aria-hidden="true" tabindex="-1"></a>render_markdown(<span class="ss">f"</span><span class="sc">{</span>ontology_response<span class="sc">.</span>text<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display cell-output-markdown">
<section id="source-2" class="level3">
<h3 class="anchored" data-anchor-id="source-2">Source</h3>
<p>Patient question describing symptoms of dyspnea on exertion, orthopnea, lower extremity edema, and weight gain in the setting of diabetes and hypertension.</p>
</section>
<section id="impression-2" class="level3">
<h3 class="anchored" data-anchor-id="impression-2">Impression</h3>
<p>🟡 <strong>Concerning for heart failure exacerbation</strong>. Patient with history of <code>diabetes</code> and <code>hypertension</code> presents with acute onset dyspnea on exertion, orthopnea, lower extremity edema, and weight gain, which are concerning for acute decompensated heart failure.</p>
</section>
<section id="supporting-evidence-2" class="level3">
<h3 class="anchored" data-anchor-id="supporting-evidence-2">Supporting Evidence</h3>
<ul>
<li><strong>Dyspnea on exertion:</strong> Difficulty breathing with movement. Mapped to <code>dyspnea on exertion</code>.
<ul>
<li>SNOMEDCT_US: 267041004</li>
<li>UMLS: C0231807</li>
</ul></li>
<li><strong>Orthopnea:</strong> Needs extra pillows to sleep. Mapped to <code>orthopnea</code>.
<ul>
<li>SNOMEDCT_US: 36748007</li>
<li>UMLS: C0085619</li>
</ul></li>
<li><strong>Lower Extremity Edema:</strong> Swelling in legs. Mapped to <code>pedal edema</code>.
<ul>
<li>SNOMEDCT_US: 102572006</li>
<li>UMLS: C0235886, C0239340</li>
</ul></li>
<li><strong>Weight gain:</strong> 4 pound weight gain. Mapped to <code>weight gain</code>, <code>increased body weight</code>.
<ul>
<li>SNOMEDCT_US: 286680006, 414916001</li>
<li>UMLS: C0043094, C1262477, C2609868</li>
</ul></li>
<li><strong>Diabetes:</strong> History of diabetes. Mapped to <code>diabetes mellitus</code>.</li>
<li><strong>Hypertension:</strong> History of hypertension. Mapped to <code>hypertension</code></li>
</ul>
<table class="table table-sm table-striped small">
<thead>
<tr class="header">
<th>STRING</th>
<th>TERMINOLOGY</th>
<th>CODE</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Dyspnea on exertion</td>
<td>SNOMEDCT_US</td>
<td>267041004</td>
</tr>
<tr class="even">
<td>Orthopnea</td>
<td>SNOMEDCT_US</td>
<td>36748007</td>
</tr>
<tr class="odd">
<td>Pedal edema</td>
<td>SNOMEDCT_US</td>
<td>102572006</td>
</tr>
<tr class="even">
<td>Weight gain</td>
<td>SNOMEDCT_US</td>
<td>286680006</td>
</tr>
<tr class="odd">
<td>Increased body weight</td>
<td>SNOMEDCT_US</td>
<td>414916001</td>
</tr>
<tr class="even">
<td>Dyspnea on exertion</td>
<td>UMLS</td>
<td>C0231807</td>
</tr>
<tr class="odd">
<td>Orthopnea</td>
<td>UMLS</td>
<td>C0085619</td>
</tr>
<tr class="even">
<td>Pedal edema</td>
<td>UMLS</td>
<td>C0235886, C0239340</td>
</tr>
<tr class="odd">
<td>Weight gain</td>
<td>UMLS</td>
<td>C0043094</td>
</tr>
<tr class="even">
<td>Increased body weight</td>
<td>UMLS</td>
<td>C1262477, C2609868</td>
</tr>
</tbody>
</table>
</section>
<section id="relevant-relationships-2" class="level3">
<h3 class="anchored" data-anchor-id="relevant-relationships-2">Relevant Relationships</h3>
<ul>
<li><code>Dyspnea on exertion</code>, <code>orthopnea</code>, and <code>pedal edema</code> are all associated with <code>heart failure</code>.</li>
<li><code>Diabetes mellitus</code> and <code>hypertension</code> are risk factors for <code>cardiovascular disease</code>, including <code>heart failure</code>.</li>
<li><code>Increased body weight</code> can exacerbate heart failure symptoms.</li>
</ul>
</section>
</div>
</div>
</section>
<section id="example-doctor-note" class="level2">
<h2 class="anchored" data-anchor-id="example-doctor-note">🎯🎯 Example Doctor note</h2>
<div id="cell-90" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T10:04:36.507569Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T10:04:36.507219Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T10:05:51.495614Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T10:05:51.494669Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T10:04:36.507538Z&quot;}" data-trusted="true" data-execution_count="64">
<div class="sourceCode cell-code" id="cb66"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb66-1"><a href="#cb66-1" aria-hidden="true" tabindex="-1"></a>user_input <span class="op">=</span> <span class="st">"""</span></span>
<span id="cb66-2"><a href="#cb66-2" aria-hidden="true" tabindex="-1"></a><span class="st">    42F presents with new-onset pancytopenia (WBC 2.1, Hgb 8.2, Plt 95) and hepatosplenomegaly, found to have t(9;22) BCR-ABL1+ with 92% blasts on bone marrow biopsy c/w CML in blast crisis.</span></span>
<span id="cb66-3"><a href="#cb66-3" aria-hidden="true" tabindex="-1"></a><span class="st">    S/p TAVR for severe AS (mean gradient 52mmHg, AVA 0.8cm²), now with progressive BiV failure (LVEF 25%, elevated JVP to 12cm) and stage III CKD (GFR 42) requiring optimization of GDMT.</span></span>
<span id="cb66-4"><a href="#cb66-4" aria-hidden="true" tabindex="-1"></a><span class="st">"""</span></span>
<span id="cb66-5"><a href="#cb66-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb66-6"><a href="#cb66-6" aria-hidden="true" tabindex="-1"></a>model_cached <span class="op">=</span> genai.GenerativeModel.from_cached_content(cached_content<span class="op">=</span>cache_all)</span>
<span id="cb66-7"><a href="#cb66-7" aria-hidden="true" tabindex="-1"></a>ontology_response <span class="op">=</span> model_cached.generate_content(</span>
<span id="cb66-8"><a href="#cb66-8" aria-hidden="true" tabindex="-1"></a>   <span class="ss">f"""Using ONLY the cached ontology relationships, provide a structured analysis report for the following input: </span><span class="sc">{</span>user_input<span class="sc">}</span></span>
<span id="cb66-9"><a href="#cb66-9" aria-hidden="true" tabindex="-1"></a><span class="ss">    </span></span>
<span id="cb66-10"><a href="#cb66-10" aria-hidden="true" tabindex="-1"></a><span class="ss">    </span><span class="sc">{</span>BASE_INSTRUCTIONS<span class="sc">}</span></span>
<span id="cb66-11"><a href="#cb66-11" aria-hidden="true" tabindex="-1"></a><span class="ss">    """</span></span>
<span id="cb66-12"><a href="#cb66-12" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb66-13"><a href="#cb66-13" aria-hidden="true" tabindex="-1"></a>render_markdown(<span class="ss">f"</span><span class="sc">{</span>ontology_response<span class="sc">.</span>text<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display cell-output-markdown">
<section id="source-3" class="level3">
<h3 class="anchored" data-anchor-id="source-3">Source</h3>
<p>Clinical text describing a 42-year-old female with CML in blast crisis and biventricular heart failure.</p>
</section>
<section id="impression-3" class="level3">
<h3 class="anchored" data-anchor-id="impression-3">Impression</h3>
<p>🔴 <strong>Chronic Myeloid Leukemia (CML) in blast crisis</strong> with pancytopenia and hepatosplenomegaly. 🔴 <strong>Severe biventricular heart failure</strong> (BiV failure) with reduced ejection fraction (LVEF 25%) and elevated jugular venous pressure (JVP). 🟡 <strong>Stage III chronic kidney disease (CKD)</strong> (GFR 42).</p>
</section>
<section id="supporting-evidence-3" class="level3">
<h3 class="anchored" data-anchor-id="supporting-evidence-3">Supporting Evidence</h3>
<ul>
<li><strong>Chronic Myeloid Leukemia (CML)</strong>
<ul>
<li><code>t(9;22)</code> and <code>BCR-ABL1+</code> genetic markers confirm diagnosis.</li>
<li>92% blasts on bone marrow biopsy indicates blast crisis.</li>
<li>SNOMEDCT_US: 154592009, 92818009, 63364005</li>
<li>UMLS: C0023473</li>
<li>Supporting details: “…t(9;22) BCR-ABL1+ with 92% blasts on bone marrow biopsy c/w CML in blast crisis.”</li>
</ul></li>
<li><strong>Pancytopenia:</strong>
<ul>
<li>SNOMEDCT_US: 127034005, 300980002</li>
<li>UMLS: C0030312</li>
<li>Supporting details: “new-onset pancytopenia (WBC 2.1, Hgb 8.2, Plt 95)”</li>
</ul></li>
<li><strong>Hepatosplenomegaly:</strong>
<ul>
<li>SNOMEDCT_US: 36760000</li>
<li>UMLS: C0019214</li>
<li>Supporting details: “…hepatosplenomegaly…”</li>
</ul></li>
<li><strong>Biventricular Heart Failure</strong>:
<ul>
<li>LVEF 25% (severely reduced)</li>
<li>SNOMEDCT_US: 42343007, 84114007, 195108009, 155374007, 274096000, 194887007, 128404006</li>
<li>ICD10CM: I50.9</li>
<li>UMLS: C0018802, C0018801</li>
<li>Elevated JVP (12 cm)</li>
<li>UMLS: C0520861</li>
<li>Supporting details: “…progressive BiV failure (LVEF 25%, elevated JVP to 12cm)…”</li>
</ul></li>
<li><strong>Severe Aortic Stenosis (AS)</strong>: Status post TAVR.
<ul>
<li>Significant mean gradient (52 mmHg) and reduced AVA (0.8 cm²) suggest prior severe AS.</li>
<li>SNOMEDCT_US: 60573004, 204368006, 250978003, 250915007, 82458004, 81817003</li>
<li>ICD10CM: I06.0, I35.0</li>
<li>UMLS: C0003507</li>
</ul></li>
<li><strong>Stage III Chronic Kidney Disease (CKD)</strong>:
<ul>
<li>SNOMEDCT_US: 431857002, 155856009, 431855005, 709044004, 42399005, 70730006</li>
<li>ICD10CM: I12.9, I13.10, N18.3</li>
<li>UMLS: C0022661, C2317473, C2316401</li>
<li>GFR 42 mL/min.</li>
<li>Supporting details: “…stage III CKD (GFR 42)…”</li>
</ul></li>
</ul>
<table class="table table-sm table-striped small">
<colgroup>
<col style="width: 33%">
<col style="width: 33%">
<col style="width: 33%">
</colgroup>
<thead>
<tr class="header">
<th style="text-align: left;">STRING</th>
<th style="text-align: left;">TERMINOLOGY</th>
<th style="text-align: left;">CODE</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td style="text-align: left;">Chronic Myeloid Leukemia</td>
<td style="text-align: left;">SNOMEDCT_US</td>
<td style="text-align: left;">63364005</td>
</tr>
<tr class="even">
<td style="text-align: left;">Chronic myeloid leukemia, BCR-ABL1 positive</td>
<td style="text-align: left;">SNOMEDCT_US</td>
<td style="text-align: left;">154592009</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Chronic myelogenous leukemia</td>
<td style="text-align: left;">SNOMEDCT_US</td>
<td style="text-align: left;">92818009</td>
</tr>
<tr class="even">
<td style="text-align: left;">Chronic Myeloid Leukemia</td>
<td style="text-align: left;">UMLS</td>
<td style="text-align: left;">C0023473</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Pancytopenia</td>
<td style="text-align: left;">SNOMEDCT_US</td>
<td style="text-align: left;">127034005</td>
</tr>
<tr class="even">
<td style="text-align: left;">Pancytopenia</td>
<td style="text-align: left;">UMLS</td>
<td style="text-align: left;">C0030312</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Hepatosplenomegaly</td>
<td style="text-align: left;">SNOMEDCT_US</td>
<td style="text-align: left;">36760000</td>
</tr>
<tr class="even">
<td style="text-align: left;">Hepatosplenomegaly</td>
<td style="text-align: left;">UMLS</td>
<td style="text-align: left;">C0019214</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Congestive heart failure</td>
<td style="text-align: left;">SNOMEDCT_US</td>
<td style="text-align: left;">155374007, 195108009, 42343007</td>
</tr>
<tr class="even">
<td style="text-align: left;">Congestive heart failure</td>
<td style="text-align: left;">UMLS</td>
<td style="text-align: left;">C0018801, C0018802</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Biventricular heart failure</td>
<td style="text-align: left;">ICD10CM</td>
<td style="text-align: left;">I50.9</td>
</tr>
<tr class="even">
<td style="text-align: left;">Elevated jugular venous pressure</td>
<td style="text-align: left;">UMLS</td>
<td style="text-align: left;">C0520861</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Aortic stenosis</td>
<td style="text-align: left;">SNOMEDCT_US</td>
<td style="text-align: left;">60573004</td>
</tr>
<tr class="even">
<td style="text-align: left;">Aortic valve stenosis</td>
<td style="text-align: left;">UMLS</td>
<td style="text-align: left;">C0003507</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Chronic kidney disease</td>
<td style="text-align: left;">SNOMEDCT_US</td>
<td style="text-align: left;">155856009</td>
</tr>
<tr class="even">
<td style="text-align: left;">Chronic kidney disease</td>
<td style="text-align: left;">UMLS</td>
<td style="text-align: left;">C0022661</td>
</tr>
<tr class="odd">
<td style="text-align: left;">Stage 3 chronic kidney disease</td>
<td style="text-align: left;">UMLS, SNOMEDCT_US</td>
<td style="text-align: left;">C2317473, 431857002</td>
</tr>
</tbody>
</table>
</section>
<section id="relevant-relationships-3" class="level3">
<h3 class="anchored" data-anchor-id="relevant-relationships-3">Relevant Relationships</h3>
<ul>
<li><strong>CML:</strong> <code>t(9;22)</code> reciprocal translocation creates <code>BCR-ABL1</code> fusion gene, driving leukemic transformation of <code>hematopoietic cells</code>. Blast crisis phase is characterized by increased <code>myeloblasts</code>.</li>
<li><strong>Heart Failure:</strong> <code>Aortic Stenosis</code> can lead to left ventricular hypertrophy and subsequent <code>heart failure</code>. TAVR intervention aims to alleviate the stenosis, but progression to biventricular failure is observed in this case. Elevated JVP indicates right heart failure.</li>
<li><strong>CKD:</strong> Can exacerbate <code>heart failure</code> through volume overload and contribute to <code>pancytopenia</code> through reduced erythropoietin production. Optimal GDMT management is crucial to address the interconnectedness of these conditions. Requires ongoing monitoring given these relationships.</li>
</ul>
</section>
</div>
</div>
</section>
<section id="image-to-concepts-dermatology" class="level2">
<h2 class="anchored" data-anchor-id="image-to-concepts-dermatology">🤯🤯 Image-to-concepts: Dermatology</h2>
<div id="cell-92" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T10:45:21.392469Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T10:45:21.392080Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T10:45:21.414169Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T10:45:21.413203Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T10:45:21.392434Z&quot;}" data-trusted="true" data-execution_count="65">
<div class="sourceCode cell-code" id="cb67"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb67-1"><a href="#cb67-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> IPython.display <span class="im">import</span> Image</span>
<span id="cb67-2"><a href="#cb67-2" aria-hidden="true" tabindex="-1"></a>IMAGE_FILEPATH <span class="op">=</span> <span class="st">'data/1730706850076.jpeg'</span></span>
<span id="cb67-3"><a href="#cb67-3" aria-hidden="true" tabindex="-1"></a>Image(filename<span class="op">=</span>IMAGE_FILEPATH, width<span class="op">=</span><span class="dv">400</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="65">
<div>
<figure class="figure">
<p><img src="ai-ontology-entity-linking-for-clinical-content_files/figure-html/cell-42-output-1.jpeg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-93" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T10:46:59.405560Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T10:46:59.405188Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T10:47:44.052752Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T10:47:44.051814Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T10:46:59.405527Z&quot;}" data-trusted="true" data-execution_count="66">
<div class="sourceCode cell-code" id="cb68"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb68-1"><a href="#cb68-1" aria-hidden="true" tabindex="-1"></a>image_model <span class="op">=</span> genai.GenerativeModel(model_name<span class="op">=</span><span class="st">"gemini-1.5-pro-latest"</span>)</span>
<span id="cb68-2"><a href="#cb68-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-3"><a href="#cb68-3" aria-hidden="true" tabindex="-1"></a>sample_file <span class="op">=</span> genai.upload_file(</span>
<span id="cb68-4"><a href="#cb68-4" aria-hidden="true" tabindex="-1"></a>    path<span class="op">=</span><span class="st">'data/1730706850076.jpeg'</span>, </span>
<span id="cb68-5"><a href="#cb68-5" aria-hidden="true" tabindex="-1"></a>    display_name<span class="op">=</span><span class="st">"hiking-rash"</span></span>
<span id="cb68-6"><a href="#cb68-6" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb68-7"><a href="#cb68-7" aria-hidden="true" tabindex="-1"></a>uploaded_file <span class="op">=</span> genai.get_file(sample_file.name)</span>
<span id="cb68-8"><a href="#cb68-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-9"><a href="#cb68-9" aria-hidden="true" tabindex="-1"></a><span class="co"># Step 1: Multimodal model to "describe" the image</span></span>
<span id="cb68-10"><a href="#cb68-10" aria-hidden="true" tabindex="-1"></a>image_response <span class="op">=</span> image_model.generate_content(</span>
<span id="cb68-11"><a href="#cb68-11" aria-hidden="true" tabindex="-1"></a>    [</span>
<span id="cb68-12"><a href="#cb68-12" aria-hidden="true" tabindex="-1"></a>        uploaded_file,</span>
<span id="cb68-13"><a href="#cb68-13" aria-hidden="true" tabindex="-1"></a>        <span class="st">"""Describe this medical image with precise clinical observations.</span></span>
<span id="cb68-14"><a href="#cb68-14" aria-hidden="true" tabindex="-1"></a><span class="st">        REQUIRED OBSERVATIONS:</span></span>
<span id="cb68-15"><a href="#cb68-15" aria-hidden="true" tabindex="-1"></a><span class="st">        ### Visual Elements</span></span>
<span id="cb68-16"><a href="#cb68-16" aria-hidden="true" tabindex="-1"></a><span class="st">        - Imaging modality/view</span></span>
<span id="cb68-17"><a href="#cb68-17" aria-hidden="true" tabindex="-1"></a><span class="st">        - Key anatomical structures</span></span>
<span id="cb68-18"><a href="#cb68-18" aria-hidden="true" tabindex="-1"></a><span class="st">        - Notable abnormalities</span></span>
<span id="cb68-19"><a href="#cb68-19" aria-hidden="true" tabindex="-1"></a><span class="st">        - Quality/technical factors</span></span>
<span id="cb68-20"><a href="#cb68-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-21"><a href="#cb68-21" aria-hidden="true" tabindex="-1"></a><span class="st">        FORMAT:</span></span>
<span id="cb68-22"><a href="#cb68-22" aria-hidden="true" tabindex="-1"></a><span class="st">        - Clear, clinical language</span></span>
<span id="cb68-23"><a href="#cb68-23" aria-hidden="true" tabindex="-1"></a><span class="st">        - Anatomical terminology</span></span>
<span id="cb68-24"><a href="#cb68-24" aria-hidden="true" tabindex="-1"></a><span class="st">        - Systematic organization</span></span>
<span id="cb68-25"><a href="#cb68-25" aria-hidden="true" tabindex="-1"></a><span class="st">        - Avoid interpretation</span></span>
<span id="cb68-26"><a href="#cb68-26" aria-hidden="true" tabindex="-1"></a><span class="st">        - Focus on objective findings"""</span>,</span>
<span id="cb68-27"><a href="#cb68-27" aria-hidden="true" tabindex="-1"></a>    ],</span>
<span id="cb68-28"><a href="#cb68-28" aria-hidden="true" tabindex="-1"></a>    request_options<span class="op">=</span>{<span class="st">"timeout"</span>: <span class="dv">600</span>},</span>
<span id="cb68-29"><a href="#cb68-29" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb68-30"><a href="#cb68-30" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb68-31"><a href="#cb68-31" aria-hidden="true" tabindex="-1"></a><span class="co"># Step 2: Use the description to annotate</span></span>
<span id="cb68-32"><a href="#cb68-32" aria-hidden="true" tabindex="-1"></a>model_cached <span class="op">=</span> genai.GenerativeModel.from_cached_content(cached_content<span class="op">=</span>cache_all)</span>
<span id="cb68-33"><a href="#cb68-33" aria-hidden="true" tabindex="-1"></a>ontology_response <span class="op">=</span> model_cached.generate_content(</span>
<span id="cb68-34"><a href="#cb68-34" aria-hidden="true" tabindex="-1"></a>   <span class="ss">f"""Using ONLY the cached ontology relationships, provide a structured analysis report for the following input: </span><span class="sc">{</span>image_response<span class="sc">.</span>text<span class="sc">}</span></span>
<span id="cb68-35"><a href="#cb68-35" aria-hidden="true" tabindex="-1"></a><span class="ss">    </span></span>
<span id="cb68-36"><a href="#cb68-36" aria-hidden="true" tabindex="-1"></a><span class="ss">    </span><span class="sc">{</span>BASE_INSTRUCTIONS<span class="sc">}</span></span>
<span id="cb68-37"><a href="#cb68-37" aria-hidden="true" tabindex="-1"></a><span class="ss">    """</span></span>
<span id="cb68-38"><a href="#cb68-38" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb68-39"><a href="#cb68-39" aria-hidden="true" tabindex="-1"></a>render_markdown(<span class="ss">f"</span><span class="sc">{</span>ontology_response<span class="sc">.</span>text<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display cell-output-markdown">
<section id="source-4" class="level3">
<h3 class="anchored" data-anchor-id="source-4">Source</h3>
<p>Photo of bilateral lower legs showing an erythematous maculopapular rash.</p>
</section>
<section id="impression-4" class="level3">
<h3 class="anchored" data-anchor-id="impression-4">Impression</h3>
<p>Bilateral erythematous maculopapular rash, possibly <code>dermatitis</code> or <code>exanthem</code>. No evidence of blistering or epidermal disruption.&nbsp;</p>
</section>
<section id="supporting-evidence-4" class="level3">
<h3 class="anchored" data-anchor-id="supporting-evidence-4">Supporting Evidence</h3>
<ul>
<li><strong>Rash</strong>:
<ul>
<li>Mapped to <code>skin rash</code></li>
<li>SNOMEDCT_US: <a href="https://browser.ihtsdotools.org/?perspective=full&amp;conceptId1=271807003&amp;edition=MAIN/SNOMEDCT-US/2024-09-01&amp;release=&amp;languages=en">271807003</a></li>
<li>UMLS: <a href="https://uts.nlm.nih.gov/uts/umls/searchResults?searchString=C0015230">C0015230</a></li>
<li>Described as “erythematous, maculopapular” affecting posterior lower legs bilaterally, sparing ankles and heels.</li>
</ul></li>
<li><strong>Erythema</strong>:
<ul>
<li>Mapped to <code>erythema</code></li>
<li>SNOMEDCT_US: <a href="https://browser.ihtsdotools.org/?perspective=full&amp;conceptId1=238810007&amp;edition=MAIN/SNOMEDCT-US/2024-09-01&amp;release=&amp;languages=en">238810007</a>, <a href="https://browser.ihtsdotools.org/?perspective=full&amp;conceptId1=327422003&amp;edition=MAIN/SNOMEDCT-US/2024-09-01&amp;release=&amp;languages=en">327422003</a></li>
<li>UMLS: <a href="https://uts.nlm.nih.gov/uts/umls/searchResults?searchString=C0016382">C0016382</a>, <a href="https://uts.nlm.nih.gov/uts/umls/searchResults?searchString=C0041834">C0041834</a></li>
<li>Noted as a component of the rash.</li>
</ul></li>
<li><strong>Macules and Papules</strong>:
<ul>
<li>Macules mapped to <code>macule</code></li>
<li>Papules mapped to <code>papule</code></li>
<li>Noted as components of the rash (“small, discrete macules and slightly raised papules”).</li>
</ul></li>
<li><strong>No Blistering</strong>:
<ul>
<li>Mapped to <code>blister</code>❌ (“no visible blistering”)</li>
</ul></li>
<li><strong>No Weeping</strong>: (“no visible weeping”)</li>
<li><strong>No Crusting</strong>: (“no visible crusting”)</li>
</ul>
<table class="table table-sm table-striped small">
<thead>
<tr class="header">
<th>STRING</th>
<th>TERMINOLOGY</th>
<th>CODE</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Skin rash</td>
<td>SNOMEDCT_US</td>
<td>271807003</td>
</tr>
<tr class="even">
<td>Skin rash</td>
<td>UMLS</td>
<td>C0015230</td>
</tr>
<tr class="odd">
<td>Erythema</td>
<td>SNOMEDCT_US</td>
<td>238810007</td>
</tr>
<tr class="even">
<td>Erythema</td>
<td>SNOMEDCT_US</td>
<td>327422003</td>
</tr>
<tr class="odd">
<td>Erythema</td>
<td>UMLS</td>
<td>C0016382</td>
</tr>
<tr class="even">
<td>Erythema</td>
<td>UMLS</td>
<td>C0041834</td>
</tr>
</tbody>
</table>
</section>
<section id="relevant-relationships-4" class="level3">
<h3 class="anchored" data-anchor-id="relevant-relationships-4">Relevant Relationships</h3>
<ul>
<li><code>Skin rash</code> can be a manifestation of <code>dermatitis</code> or <code>exanthem</code>.</li>
<li><code>Erythema</code> is a visual characteristic of skin often associated with <code>inflammatory abnormality of the skin</code>.</li>
<li><code>Macule</code> and <code>papule</code> are specific <code>morphological abnormality of the skin</code>.</li>
<li>The absence of <code>blister</code> suggests that the epidermal layer is intact.&nbsp;</li>
</ul>
</section>
</div>
</div>
</section>
<section id="image-to-concepts-radiology" class="level2">
<h2 class="anchored" data-anchor-id="image-to-concepts-radiology">🤯🤯 Image-to-concepts: Radiology</h2>
<div id="cell-95" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T10:46:07.181794Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T10:46:07.181462Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T10:46:07.195173Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T10:46:07.194106Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T10:46:07.181738Z&quot;}" data-trusted="true" data-execution_count="67">
<div class="sourceCode cell-code" id="cb69"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb69-1"><a href="#cb69-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> IPython.display <span class="im">import</span> Image</span>
<span id="cb69-2"><a href="#cb69-2" aria-hidden="true" tabindex="-1"></a>IMAGE_FILEPATH <span class="op">=</span> <span class="st">'data/consolidation-rml.jpg'</span></span>
<span id="cb69-3"><a href="#cb69-3" aria-hidden="true" tabindex="-1"></a>Image(filename<span class="op">=</span>IMAGE_FILEPATH, width<span class="op">=</span><span class="dv">400</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="67">
<div>
<figure class="figure">
<p><img src="ai-ontology-entity-linking-for-clinical-content_files/figure-html/cell-44-output-1.jpeg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-96" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T10:46:07.196834Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T10:46:07.196503Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T10:46:46.569594Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T10:46:46.568559Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T10:46:07.196799Z&quot;}" data-trusted="true" data-execution_count="68">
<div class="sourceCode cell-code" id="cb70"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb70-1"><a href="#cb70-1" aria-hidden="true" tabindex="-1"></a>image_model <span class="op">=</span> genai.GenerativeModel(model_name<span class="op">=</span><span class="st">"gemini-1.5-pro-latest"</span>)</span>
<span id="cb70-2"><a href="#cb70-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-3"><a href="#cb70-3" aria-hidden="true" tabindex="-1"></a>sample_file <span class="op">=</span> genai.upload_file(</span>
<span id="cb70-4"><a href="#cb70-4" aria-hidden="true" tabindex="-1"></a>    path<span class="op">=</span><span class="st">'data/consolidation-rml.jpg'</span>, </span>
<span id="cb70-5"><a href="#cb70-5" aria-hidden="true" tabindex="-1"></a>    display_name<span class="op">=</span><span class="st">"radiology-consolidation"</span>)</span>
<span id="cb70-6"><a href="#cb70-6" aria-hidden="true" tabindex="-1"></a>uploaded_file <span class="op">=</span> genai.get_file(sample_file.name)</span>
<span id="cb70-7"><a href="#cb70-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-8"><a href="#cb70-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Step 1: Multimodal model to "describe" the image</span></span>
<span id="cb70-9"><a href="#cb70-9" aria-hidden="true" tabindex="-1"></a>image_response <span class="op">=</span> image_model.generate_content(</span>
<span id="cb70-10"><a href="#cb70-10" aria-hidden="true" tabindex="-1"></a>    [</span>
<span id="cb70-11"><a href="#cb70-11" aria-hidden="true" tabindex="-1"></a>        uploaded_file,</span>
<span id="cb70-12"><a href="#cb70-12" aria-hidden="true" tabindex="-1"></a>        <span class="st">"""Describe this medical image with precise clinical observations.</span></span>
<span id="cb70-13"><a href="#cb70-13" aria-hidden="true" tabindex="-1"></a><span class="st">        REQUIRED OBSERVATIONS:</span></span>
<span id="cb70-14"><a href="#cb70-14" aria-hidden="true" tabindex="-1"></a><span class="st">        ### Visual Elements</span></span>
<span id="cb70-15"><a href="#cb70-15" aria-hidden="true" tabindex="-1"></a><span class="st">        - Imaging modality/view</span></span>
<span id="cb70-16"><a href="#cb70-16" aria-hidden="true" tabindex="-1"></a><span class="st">        - Key anatomical structures</span></span>
<span id="cb70-17"><a href="#cb70-17" aria-hidden="true" tabindex="-1"></a><span class="st">        - Notable abnormalities</span></span>
<span id="cb70-18"><a href="#cb70-18" aria-hidden="true" tabindex="-1"></a><span class="st">        - Quality/technical factors</span></span>
<span id="cb70-19"><a href="#cb70-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-20"><a href="#cb70-20" aria-hidden="true" tabindex="-1"></a><span class="st">        FORMAT:</span></span>
<span id="cb70-21"><a href="#cb70-21" aria-hidden="true" tabindex="-1"></a><span class="st">        - Clear, clinical language</span></span>
<span id="cb70-22"><a href="#cb70-22" aria-hidden="true" tabindex="-1"></a><span class="st">        - Anatomical terminology</span></span>
<span id="cb70-23"><a href="#cb70-23" aria-hidden="true" tabindex="-1"></a><span class="st">        - Systematic organization</span></span>
<span id="cb70-24"><a href="#cb70-24" aria-hidden="true" tabindex="-1"></a><span class="st">        - Avoid interpretation</span></span>
<span id="cb70-25"><a href="#cb70-25" aria-hidden="true" tabindex="-1"></a><span class="st">        - Focus on objective findings"""</span>,</span>
<span id="cb70-26"><a href="#cb70-26" aria-hidden="true" tabindex="-1"></a>    ],</span>
<span id="cb70-27"><a href="#cb70-27" aria-hidden="true" tabindex="-1"></a>    request_options<span class="op">=</span>{<span class="st">"timeout"</span>: <span class="dv">600</span>},</span>
<span id="cb70-28"><a href="#cb70-28" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb70-29"><a href="#cb70-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb70-30"><a href="#cb70-30" aria-hidden="true" tabindex="-1"></a><span class="co"># Step 2: Use the description to annotate</span></span>
<span id="cb70-31"><a href="#cb70-31" aria-hidden="true" tabindex="-1"></a>model_cached <span class="op">=</span> genai.GenerativeModel.from_cached_content(cached_content<span class="op">=</span>cache_all)</span>
<span id="cb70-32"><a href="#cb70-32" aria-hidden="true" tabindex="-1"></a>ontology_response <span class="op">=</span> model_cached.generate_content(</span>
<span id="cb70-33"><a href="#cb70-33" aria-hidden="true" tabindex="-1"></a>    <span class="ss">f"""Using ONLY the cached ontology relationships, provide a structured analysis report for the following input: </span><span class="sc">{</span>image_response<span class="sc">.</span>text<span class="sc">}</span></span>
<span id="cb70-34"><a href="#cb70-34" aria-hidden="true" tabindex="-1"></a><span class="ss">    </span></span>
<span id="cb70-35"><a href="#cb70-35" aria-hidden="true" tabindex="-1"></a><span class="ss">    </span><span class="sc">{</span>BASE_INSTRUCTIONS<span class="sc">}</span></span>
<span id="cb70-36"><a href="#cb70-36" aria-hidden="true" tabindex="-1"></a><span class="ss">    """</span>,</span>
<span id="cb70-37"><a href="#cb70-37" aria-hidden="true" tabindex="-1"></a>    request_options<span class="op">=</span>{<span class="st">"timeout"</span>: <span class="dv">600</span>}</span>
<span id="cb70-38"><a href="#cb70-38" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb70-39"><a href="#cb70-39" aria-hidden="true" tabindex="-1"></a>render_markdown(<span class="ss">f"</span><span class="sc">{</span>ontology_response<span class="sc">.</span>text<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display cell-output-markdown">
<section id="source-5" class="level3">
<h3 class="anchored" data-anchor-id="source-5">Source</h3>
<p>Frontal chest radiograph demonstrating diffuse, bilateral interstitial and alveolar opacities.</p>
</section>
<section id="impression-5" class="level3">
<h3 class="anchored" data-anchor-id="impression-5">Impression</h3>
<p>Diffuse <strong>lung opacities</strong>, possibly representing pneumonia. No pneumothorax or abnormalities of the cardiomediastinal silhouette or diaphragm. Increased density in the lower lung fields due to breast tissue.</p>
</section>
<section id="supporting-evidence-5" class="level3">
<h3 class="anchored" data-anchor-id="supporting-evidence-5">Supporting Evidence</h3>
<ul>
<li><strong>Bilateral interstitial opacity</strong>:
<ul>
<li><code>pulmonary infiltrates</code>
<ul>
<li>UMLS: <a href="https://uts.nlm.nih.gov/uts/umls/searchResults?searchString=C0235896">C0235896</a></li>
</ul></li>
<li>Supporting detail: The lungs demonstrate diffuse, bilateral interstitial … opacities</li>
</ul></li>
<li><strong>Bilateral alveolar opacity</strong>:
<ul>
<li>SNOMEDCT_US: <a href="https://browser.ihtsdotools.org/?perspective=full&amp;conceptId1=10501004&amp;edition=MAIN/SNOMEDCT-US/2024-09-01&amp;release=&amp;languages=en">10501004</a> -UMLS: <a href="https://uts.nlm.nih.gov/uts/umls/searchResults?searchString=C0034050">C0034050</a></li>
<li>Supporting detail: … bilateral … alveolar opacities</li>
</ul></li>
<li><strong>Basilar consolidation</strong>:
<ul>
<li>Possible <code>pneumonia</code>
<ul>
<li>SNOMEDCT_US: <a href="https://browser.ihtsdotools.org/?perspective=full&amp;conceptId1=233604007&amp;edition=MAIN/SNOMEDCT-US/2024-09-01&amp;release=&amp;languages=en">233604007</a></li>
<li>UMLS: <a href="https://uts.nlm.nih.gov/uts/umls/searchResults?searchString=C0032285">C0032285</a></li>
</ul></li>
<li>Supporting detail: … possible basilar consolidation in the left lung.</li>
</ul></li>
<li>❌ <strong>Pneumothorax</strong>:
<ul>
<li><code>pneumothorax</code>
<ul>
<li>SNOMEDCT_US: <a href="https://browser.ihtsdotools.org/?perspective=full&amp;conceptId1=196103008&amp;edition=MAIN/SNOMEDCT-US/2024-09-01&amp;release=&amp;languages=en">196103008</a></li>
<li>UMLS: <a href="https://uts.nlm.nih.gov/uts/umls/searchResults?searchString=C0029850">C0029850</a></li>
<li>Supporting detail: No pneumothorax is seen.</li>
</ul></li>
</ul></li>
<li>Increased density in lower lung fields due to overlying breast tissue.</li>
</ul>
<table class="table table-sm table-striped small">
<thead>
<tr class="header">
<th>STRING</th>
<th>TERMINOLOGY</th>
<th>CODE</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>pulmonary infiltrates</td>
<td>UMLS</td>
<td>C0235896</td>
</tr>
<tr class="even">
<td>intra-alveolar opacity</td>
<td>SNOMEDCT_US</td>
<td>10501004</td>
</tr>
<tr class="odd">
<td>intra-alveolar opacity</td>
<td>UMLS</td>
<td>C0034050</td>
</tr>
<tr class="even">
<td>pneumonia</td>
<td>SNOMEDCT_US</td>
<td>233604007</td>
</tr>
<tr class="odd">
<td>pneumonia</td>
<td>UMLS</td>
<td>C0032285</td>
</tr>
<tr class="even">
<td>pneumothorax</td>
<td>SNOMEDCT_US</td>
<td>196103008</td>
</tr>
<tr class="odd">
<td>pneumothorax</td>
<td>UMLS</td>
<td>C0029850</td>
</tr>
</tbody>
</table>
</section>
<section id="relevant-relationships-5" class="level3">
<h3 class="anchored" data-anchor-id="relevant-relationships-5">Relevant Relationships</h3>
<ul>
<li><code>Pneumonia</code> is a <code>lung disease</code>.</li>
<li><code>Pulmonary infiltrates</code> is found in <code>pneumonia</code>.</li>
<li><code>Pneumothorax</code> is a <code>lung disease</code>.</li>
<li><code>Interstitial opacity</code> is a finding associated with various lung diseases, including <code>pneumonia</code>.</li>
<li><code>Alveolar opacity</code> is a finding associated with various lung diseases, including <code>pneumonia</code>.</li>
</ul>
</section>
</div>
</div>
<div id="cell-97" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T09:45:27.200260Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T09:45:27.199833Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T09:45:27.235512Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T09:45:27.234447Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T09:45:27.200204Z&quot;}" data-trusted="true" data-execution_count="69">
<div class="sourceCode cell-code" id="cb71"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb71-1"><a href="#cb71-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> IPython.display <span class="im">import</span> Image</span>
<span id="cb71-2"><a href="#cb71-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Image from https://en.wikipedia.org/wiki/Bell%27s_palsy</span></span>
<span id="cb71-3"><a href="#cb71-3" aria-hidden="true" tabindex="-1"></a>IMAGE_FILEPATH <span class="op">=</span> <span class="st">'data/SCR-20241201-cubw.jpeg'</span></span>
<span id="cb71-4"><a href="#cb71-4" aria-hidden="true" tabindex="-1"></a>Image(filename<span class="op">=</span>IMAGE_FILEPATH, width<span class="op">=</span><span class="dv">300</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="69">
<div>
<figure class="figure">
<p><img src="ai-ontology-entity-linking-for-clinical-content_files/figure-html/cell-46-output-1.jpeg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-98" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T10:02:20.329146Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T10:02:20.328671Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T10:03:10.813380Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T10:03:10.812150Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T10:02:20.329109Z&quot;}" data-trusted="true" data-execution_count="70">
<div class="sourceCode cell-code" id="cb72"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb72-1"><a href="#cb72-1" aria-hidden="true" tabindex="-1"></a>image_model <span class="op">=</span> genai.GenerativeModel(model_name<span class="op">=</span><span class="st">"gemini-1.5-pro-latest"</span>)</span>
<span id="cb72-2"><a href="#cb72-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-3"><a href="#cb72-3" aria-hidden="true" tabindex="-1"></a>sample_file <span class="op">=</span> genai.upload_file(</span>
<span id="cb72-4"><a href="#cb72-4" aria-hidden="true" tabindex="-1"></a>    path<span class="op">=</span><span class="st">"data/SCR-20241201-cubw.jpeg"</span>, </span>
<span id="cb72-5"><a href="#cb72-5" aria-hidden="true" tabindex="-1"></a>    display_name<span class="op">=</span><span class="st">"Bells Palsy"</span>)</span>
<span id="cb72-6"><a href="#cb72-6" aria-hidden="true" tabindex="-1"></a>uploaded_file <span class="op">=</span> genai.get_file(sample_file.name)</span>
<span id="cb72-7"><a href="#cb72-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-8"><a href="#cb72-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Step 1: Multimodal model to "describe" the image</span></span>
<span id="cb72-9"><a href="#cb72-9" aria-hidden="true" tabindex="-1"></a>image_response <span class="op">=</span> image_model.generate_content(</span>
<span id="cb72-10"><a href="#cb72-10" aria-hidden="true" tabindex="-1"></a>    [</span>
<span id="cb72-11"><a href="#cb72-11" aria-hidden="true" tabindex="-1"></a>        uploaded_file,</span>
<span id="cb72-12"><a href="#cb72-12" aria-hidden="true" tabindex="-1"></a>        <span class="st">"""Describe this medical image with precise clinical observations.</span></span>
<span id="cb72-13"><a href="#cb72-13" aria-hidden="true" tabindex="-1"></a><span class="st">        REQUIRED OBSERVATIONS:</span></span>
<span id="cb72-14"><a href="#cb72-14" aria-hidden="true" tabindex="-1"></a><span class="st">        ### Visual Elements</span></span>
<span id="cb72-15"><a href="#cb72-15" aria-hidden="true" tabindex="-1"></a><span class="st">        - Imaging modality/view</span></span>
<span id="cb72-16"><a href="#cb72-16" aria-hidden="true" tabindex="-1"></a><span class="st">        - Key anatomical structures</span></span>
<span id="cb72-17"><a href="#cb72-17" aria-hidden="true" tabindex="-1"></a><span class="st">        - Notable abnormalities</span></span>
<span id="cb72-18"><a href="#cb72-18" aria-hidden="true" tabindex="-1"></a><span class="st">        - Quality/technical factors</span></span>
<span id="cb72-19"><a href="#cb72-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-20"><a href="#cb72-20" aria-hidden="true" tabindex="-1"></a><span class="st">        FORMAT:</span></span>
<span id="cb72-21"><a href="#cb72-21" aria-hidden="true" tabindex="-1"></a><span class="st">        - Clear, clinical language</span></span>
<span id="cb72-22"><a href="#cb72-22" aria-hidden="true" tabindex="-1"></a><span class="st">        - Anatomical terminology</span></span>
<span id="cb72-23"><a href="#cb72-23" aria-hidden="true" tabindex="-1"></a><span class="st">        - Systematic organization</span></span>
<span id="cb72-24"><a href="#cb72-24" aria-hidden="true" tabindex="-1"></a><span class="st">        - Avoid interpretation</span></span>
<span id="cb72-25"><a href="#cb72-25" aria-hidden="true" tabindex="-1"></a><span class="st">        - Focus on objective findings"""</span>,</span>
<span id="cb72-26"><a href="#cb72-26" aria-hidden="true" tabindex="-1"></a>    ],</span>
<span id="cb72-27"><a href="#cb72-27" aria-hidden="true" tabindex="-1"></a>    request_options<span class="op">=</span>{<span class="st">"timeout"</span>: <span class="dv">600</span>},</span>
<span id="cb72-28"><a href="#cb72-28" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb72-29"><a href="#cb72-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-30"><a href="#cb72-30" aria-hidden="true" tabindex="-1"></a><span class="co"># Step 2: Use the description to annotate</span></span>
<span id="cb72-31"><a href="#cb72-31" aria-hidden="true" tabindex="-1"></a>model_cached <span class="op">=</span> genai.GenerativeModel.from_cached_content(cached_content<span class="op">=</span>cache_all)</span>
<span id="cb72-32"><a href="#cb72-32" aria-hidden="true" tabindex="-1"></a>ontology_response <span class="op">=</span> model_cached.generate_content(</span>
<span id="cb72-33"><a href="#cb72-33" aria-hidden="true" tabindex="-1"></a>    <span class="ss">f"""Using ONLY the cached ontology relationships, provide a structured analysis report for the following input: </span><span class="sc">{</span>image_response<span class="sc">.</span>text<span class="sc">}</span></span>
<span id="cb72-34"><a href="#cb72-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb72-35"><a href="#cb72-35" aria-hidden="true" tabindex="-1"></a><span class="ss">    </span><span class="sc">{</span>BASE_INSTRUCTIONS<span class="sc">}</span></span>
<span id="cb72-36"><a href="#cb72-36" aria-hidden="true" tabindex="-1"></a><span class="ss">    """</span>,</span>
<span id="cb72-37"><a href="#cb72-37" aria-hidden="true" tabindex="-1"></a>    request_options<span class="op">=</span>{<span class="st">"timeout"</span>: <span class="dv">600</span>}</span>
<span id="cb72-38"><a href="#cb72-38" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb72-39"><a href="#cb72-39" aria-hidden="true" tabindex="-1"></a>render_markdown(<span class="ss">f"</span><span class="sc">{</span>ontology_response<span class="sc">.</span>text<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display cell-output-markdown">
<section id="source-6" class="level3">
<h3 class="anchored" data-anchor-id="source-6">Source</h3>
<p>Photo of a male’s face, frontal view, showing right-sided facial droop, erythema, and possible edema.</p>
</section>
<section id="impression-6" class="level3">
<h3 class="anchored" data-anchor-id="impression-6">Impression</h3>
<p>🟡 Right-sided facial asymmetry and erythema concerning for possible facial nerve palsy. Possible mild edema. Image resolution limits definitive diagnosis.</p>
</section>
<section id="supporting-evidence-6" class="level3">
<h3 class="anchored" data-anchor-id="supporting-evidence-6">Supporting Evidence</h3>
<ul>
<li><strong>Facial Asymmetry:</strong>
<ul>
<li>Mapped to <code>facial asymmetry</code></li>
<li>UMLS: C4022692</li>
<li>Supporting detail: “A slight droop is observed on the right side of the face, particularly around the mouth. The right side of the mouth appears lower than the left.”</li>
</ul></li>
<li><strong>Erythema:</strong>
<ul>
<li>Mapped to <code>facial erythema</code></li>
<li>SNOMEDCT_US: <a href="https://browser.ihtsdotools.org/?perspective=full&amp;conceptId1=20255002&amp;edition=MAIN/SNOMEDCT-US/2024-09-01&amp;release=&amp;languages=en">20255002</a></li>
<li>UMLS: C0005874</li>
<li>Supporting detail: “Present on the cheeks and to a lesser extent on the forehead. The redness on the right cheek appears more pronounced and extends down towards the jawline.”</li>
</ul></li>
<li><strong>Edema:</strong>
<ul>
<li>Mapped to <code>facial edema</code></li>
<li>SNOMEDCT_US: <a href="https://browser.ihtsdotools.org/?perspective=full&amp;conceptId1=445088006&amp;edition=MAIN/SNOMEDCT-US/2024-09-01&amp;release=&amp;languages=en">445088006</a></li>
<li>UMLS: C0542571</li>
<li>Supporting detail: “Possible mild swelling is noted on the right cheek, contributing to the asymmetry. It is difficult to definitively assess edema from this single image.”</li>
</ul></li>
</ul>
<table class="table table-sm table-striped small">
<thead>
<tr class="header">
<th>STRING</th>
<th>TERMINOLOGY</th>
<th>CODE</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>facial asymmetry</td>
<td>UMLS</td>
<td>C4022692</td>
</tr>
<tr class="even">
<td>facial erythema</td>
<td>SNOMEDCT_US</td>
<td><a href="https://browser.ihtsdotools.org/?perspective=full&amp;conceptId1=20255002&amp;edition=MAIN/SNOMEDCT-US/2024-09-01&amp;release=&amp;languages=en">20255002</a></td>
</tr>
<tr class="odd">
<td>facial erythema</td>
<td>UMLS</td>
<td>C0005874</td>
</tr>
<tr class="even">
<td>facial edema</td>
<td>SNOMEDCT_US</td>
<td><a href="https://browser.ihtsdotools.org/?perspective=full&amp;conceptId1=445088006&amp;edition=MAIN/SNOMEDCT-US/2024-09-01&amp;release=&amp;languages=en">445088006</a></td>
</tr>
<tr class="odd">
<td>facial edema</td>
<td>UMLS</td>
<td>C0542571</td>
</tr>
</tbody>
</table>
</section>
<section id="relevant-relationships-6" class="level3">
<h3 class="anchored" data-anchor-id="relevant-relationships-6">Relevant Relationships</h3>
<ul>
<li><code>Facial asymmetry</code> can be a finding in <code>facial nerve palsy</code> (Bell’s Palsy).</li>
<li><code>Facial erythema</code> can be associated with inflammation or infection.</li>
<li><code>Facial edema</code> often co-occurs with <code>facial asymmetry</code> in cases of facial palsy. It is also associated with inflammation and infection.</li>
</ul>
</section>
</div>
</div>
</section>
<section id="image-to-concepts-general-photo" class="level2">
<h2 class="anchored" data-anchor-id="image-to-concepts-general-photo">🤯🤯 Image-to-concepts: General Photo</h2>
<div id="cell-100" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T09:46:11.515660Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T09:46:11.515376Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T09:46:11.546370Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T09:46:11.545352Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T09:46:11.515632Z&quot;}" data-trusted="true" data-execution_count="71">
<div class="sourceCode cell-code" id="cb73"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb73-1"><a href="#cb73-1" aria-hidden="true" tabindex="-1"></a><span class="im">from</span> IPython.display <span class="im">import</span> Image</span>
<span id="cb73-2"><a href="#cb73-2" aria-hidden="true" tabindex="-1"></a>IMAGE_FILEPATH <span class="op">=</span> <span class="st">'data/SCR-20241201-crfl-2.jpeg'</span></span>
<span id="cb73-3"><a href="#cb73-3" aria-hidden="true" tabindex="-1"></a>Image(filename<span class="op">=</span>IMAGE_FILEPATH, width<span class="op">=</span><span class="dv">300</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display" data-execution_count="71">
<div>
<figure class="figure">
<p><img src="ai-ontology-entity-linking-for-clinical-content_files/figure-html/cell-48-output-1.jpeg" class="img-fluid figure-img"></p>
</figure>
</div>
</div>
</div>
<div id="cell-101" class="cell" data-execution="{&quot;iopub.execute_input&quot;:&quot;2024-12-01T10:03:10.815630Z&quot;,&quot;iopub.status.busy&quot;:&quot;2024-12-01T10:03:10.814872Z&quot;,&quot;iopub.status.idle&quot;:&quot;2024-12-01T10:04:36.505836Z&quot;,&quot;shell.execute_reply&quot;:&quot;2024-12-01T10:04:36.504827Z&quot;,&quot;shell.execute_reply.started&quot;:&quot;2024-12-01T10:03:10.815586Z&quot;}" data-trusted="true">
<div class="sourceCode cell-code" id="cb74"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb74-1"><a href="#cb74-1" aria-hidden="true" tabindex="-1"></a>image_model <span class="op">=</span> genai.GenerativeModel(model_name<span class="op">=</span><span class="st">"gemini-1.5-pro-latest"</span>)</span>
<span id="cb74-2"><a href="#cb74-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-3"><a href="#cb74-3" aria-hidden="true" tabindex="-1"></a>sample_file <span class="op">=</span> genai.upload_file(</span>
<span id="cb74-4"><a href="#cb74-4" aria-hidden="true" tabindex="-1"></a>    path<span class="op">=</span><span class="st">"data/SCR-20241201-crfl-2.jpeg"</span>, </span>
<span id="cb74-5"><a href="#cb74-5" aria-hidden="true" tabindex="-1"></a>    display_name<span class="op">=</span><span class="st">"Lebron James"</span>)</span>
<span id="cb74-6"><a href="#cb74-6" aria-hidden="true" tabindex="-1"></a>uploaded_file <span class="op">=</span> genai.get_file(sample_file.name)</span>
<span id="cb74-7"><a href="#cb74-7" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-8"><a href="#cb74-8" aria-hidden="true" tabindex="-1"></a><span class="co"># Step 1: Multimodal model to "describe" the image</span></span>
<span id="cb74-9"><a href="#cb74-9" aria-hidden="true" tabindex="-1"></a>image_response <span class="op">=</span> image_model.generate_content(</span>
<span id="cb74-10"><a href="#cb74-10" aria-hidden="true" tabindex="-1"></a>    [</span>
<span id="cb74-11"><a href="#cb74-11" aria-hidden="true" tabindex="-1"></a>        uploaded_file,</span>
<span id="cb74-12"><a href="#cb74-12" aria-hidden="true" tabindex="-1"></a>        <span class="st">"""Describe this medical image with precise clinical observations.</span></span>
<span id="cb74-13"><a href="#cb74-13" aria-hidden="true" tabindex="-1"></a><span class="st">        REQUIRED OBSERVATIONS:</span></span>
<span id="cb74-14"><a href="#cb74-14" aria-hidden="true" tabindex="-1"></a><span class="st">        ### Visual Elements</span></span>
<span id="cb74-15"><a href="#cb74-15" aria-hidden="true" tabindex="-1"></a><span class="st">        - Imaging modality/view</span></span>
<span id="cb74-16"><a href="#cb74-16" aria-hidden="true" tabindex="-1"></a><span class="st">        - Key anatomical structures</span></span>
<span id="cb74-17"><a href="#cb74-17" aria-hidden="true" tabindex="-1"></a><span class="st">        - Notable abnormalities</span></span>
<span id="cb74-18"><a href="#cb74-18" aria-hidden="true" tabindex="-1"></a><span class="st">        - Quality/technical factors</span></span>
<span id="cb74-19"><a href="#cb74-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-20"><a href="#cb74-20" aria-hidden="true" tabindex="-1"></a><span class="st">        FORMAT:</span></span>
<span id="cb74-21"><a href="#cb74-21" aria-hidden="true" tabindex="-1"></a><span class="st">        - Clear, clinical language</span></span>
<span id="cb74-22"><a href="#cb74-22" aria-hidden="true" tabindex="-1"></a><span class="st">        - Anatomical terminology</span></span>
<span id="cb74-23"><a href="#cb74-23" aria-hidden="true" tabindex="-1"></a><span class="st">        - Systematic organization</span></span>
<span id="cb74-24"><a href="#cb74-24" aria-hidden="true" tabindex="-1"></a><span class="st">        - Avoid interpretation</span></span>
<span id="cb74-25"><a href="#cb74-25" aria-hidden="true" tabindex="-1"></a><span class="st">        - Focus on objective findings"""</span>,</span>
<span id="cb74-26"><a href="#cb74-26" aria-hidden="true" tabindex="-1"></a>    ],</span>
<span id="cb74-27"><a href="#cb74-27" aria-hidden="true" tabindex="-1"></a>    request_options<span class="op">=</span>{<span class="st">"timeout"</span>: <span class="dv">600</span>},</span>
<span id="cb74-28"><a href="#cb74-28" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb74-29"><a href="#cb74-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-30"><a href="#cb74-30" aria-hidden="true" tabindex="-1"></a><span class="co"># Step 2: Use the description to annotate</span></span>
<span id="cb74-31"><a href="#cb74-31" aria-hidden="true" tabindex="-1"></a>model_cached <span class="op">=</span> genai.GenerativeModel.from_cached_content(cached_content<span class="op">=</span>cache_all)</span>
<span id="cb74-32"><a href="#cb74-32" aria-hidden="true" tabindex="-1"></a>ontology_response <span class="op">=</span> model_cached.generate_content(</span>
<span id="cb74-33"><a href="#cb74-33" aria-hidden="true" tabindex="-1"></a>    <span class="ss">f"""Using ONLY the cached ontology relationships, provide a structured analysis report for the following input: </span><span class="sc">{</span>image_response<span class="sc">.</span>text<span class="sc">}</span></span>
<span id="cb74-34"><a href="#cb74-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb74-35"><a href="#cb74-35" aria-hidden="true" tabindex="-1"></a><span class="ss">    </span><span class="sc">{</span>BASE_INSTRUCTIONS<span class="sc">}</span></span>
<span id="cb74-36"><a href="#cb74-36" aria-hidden="true" tabindex="-1"></a><span class="ss">    """</span>,</span>
<span id="cb74-37"><a href="#cb74-37" aria-hidden="true" tabindex="-1"></a>    request_options<span class="op">=</span>{<span class="st">"timeout"</span>: <span class="dv">600</span>}</span>
<span id="cb74-38"><a href="#cb74-38" aria-hidden="true" tabindex="-1"></a>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<div id="cell-102" class="cell" data-execution_count="81">
<div class="sourceCode cell-code" id="cb75"><pre class="sourceCode python code-with-copy"><code class="sourceCode python"><span id="cb75-1"><a href="#cb75-1" aria-hidden="true" tabindex="-1"></a>render_markdown(<span class="ss">f"</span><span class="sc">{</span>ontology_response<span class="sc">.</span>text<span class="sc">}</span><span class="ss">"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-display cell-output-markdown">
<section id="source-7" class="level3">
<h3 class="anchored" data-anchor-id="source-7">Source</h3>
<p>Photo of head and neck, upper extremities, thorax, and abdomen taken during a basketball game.</p>
</section>
<section id="impression-7" class="level3">
<h3 class="anchored" data-anchor-id="impression-7">Impression</h3>
<p>🟢 No apparent abnormalities of the head and neck, upper extremities, thorax, or abdomen.</p>
</section>
<section id="supporting-evidence-7" class="level3">
<h3 class="anchored" data-anchor-id="supporting-evidence-7">Supporting Evidence</h3>
<ul>
<li><strong>Head and neck:</strong> Normal, unremarkable.
<ul>
<li><code>Neck musculature</code> present and appears normal.<br>
</li>
<li><code>Beard</code> present, no abnormalities described.</li>
<li><code>Facial features</code> visualized, no abnormalities described.</li>
</ul></li>
<li><strong>Upper extremity:</strong> Normal, unremarkable.
<ul>
<li><code>Shoulders</code>, <code>arms</code>, <code>hands</code>, and <code>fingers</code> visualized, no abnormalities described.</li>
<li>Tattoos noted on right arm (not clinically significant).</li>
<li>Wristband present on right wrist (not clinically significant).</li>
</ul></li>
<li><strong>Thorax</strong>: Normal, unremarkable.
<ul>
<li><code>Chest</code> and <code>upper back</code> visualized, no abnormalities described.</li>
</ul></li>
<li><strong>Abdomen</strong>: No abnormalities mentioned.
<ul>
<li>Partially visualized.</li>
</ul></li>
<li>Quality/Technical factors: Good image quality, adequate lighting and resolution.</li>
</ul>
<table class="table table-sm table-striped small">
<thead>
<tr class="header">
<th>STRING</th>
<th>TERMINOLOGY</th>
<th>CODE</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Neck musculature</td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td>Beard</td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td>Facial features</td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td>Shoulders</td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td>Arms</td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td>Hands</td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td>Fingers</td>
<td></td>
<td></td>
</tr>
<tr class="even">
<td>Chest</td>
<td></td>
<td></td>
</tr>
<tr class="odd">
<td>Upper back</td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
</section>
<section id="relevant-relationships-7" class="level3">
<h3 class="anchored" data-anchor-id="relevant-relationships-7">Relevant Relationships</h3>
<ul>
<li><code>Neck musculature</code> is part of the <code>musculoskeletal system</code>.</li>
<li><code>Beard</code> and <code>facial features</code> are part of the <code>integumentary system</code>.</li>
<li><code>Shoulders</code>, <code>arms</code>, <code>hands</code>, and <code>fingers</code> are components of the <code>upper extremity</code>, which is part of the <code>appendicular skeleton</code>.</li>
<li><code>Chest</code> and <code>upper back</code> are parts of the <code>thorax</code>.</li>
<li>The <code>thorax</code> contains organs of the <code>cardiovascular system</code> and <code>respiratory system</code>.</li>
<li>The <code>abdomen</code> contains organs of the <code>digestive system</code>, <code>urinary system</code>, and parts of the <code>reproductive system</code>.</li>
</ul>
</section>
</div>
</div>
<hr>
</section>
</section>
<section id="final-words" class="level1">
<h1>Final Words</h1>
<p>As much as the flashy innovation doesn’t exist with this project, I find the robustness to handle almost any input related to clinical AND biomedical text impressive.</p>
<p>Given many are either: - Dismissing LLM-based entity linking as impossible task. - Need a very complex RAG-based, GraphDB based mechanism for fuzzy retrieval.</p>
<p>This route of representing triplestores as plain-English, I believe gives most bang-for-the-performance without having to commit to a tech stack commitment.</p>
<hr>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      return note.innerHTML;
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>